// @injectivelabs/wallet-turnkey/dist/cjs/strategy/turnkey/otp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TurnkeyOtpWallet = void 0;
const exceptions_1 = require("@injectivelabs/exceptions");
const consts_js_1 = require("../consts.js");
class TurnkeyOtpWallet {
    static async initEmailOTP(args) {
        const { client, iframeClient } = args;
        try {
            const targetPublicKey = iframeClient.iframePublicKey;
            if (!targetPublicKey) {
                throw new exceptions_1.WalletException(new Error('Target public key not found'));
            }
            // client.$post is undefined, resorting to this for now
            const response = await client.post(args.otpInitPath || consts_js_1.TURNKEY_OTP_INIT_PATH, {
                targetPublicKey,
                email: args.email,
                suborgId: args.subOrgId,
                invalidateExistingSessions: args.invalidateExistingSessions,
            });
            return response?.data;
        }
        catch (e) {
            throw new exceptions_1.WalletException(new Error(e.message), {
                code: exceptions_1.UnspecifiedErrorCode,
                type: exceptions_1.ErrorType.WalletError,
                contextModule: 'turnkey-init-email-otp',
            });
        }
    }
    static async confirmEmailOTP(args) {
        const { client, iframeClient, expirationSeconds } = args;
        try {
            const organizationId = args.organizationId;
            const emailOTPId = args.emailOTPId;
            const targetPublicKey = iframeClient.iframePublicKey;
            const otpVerifyPath = args.otpVerifyPath || consts_js_1.TURNKEY_OTP_VERIFY_PATH;
            if (!targetPublicKey) {
                throw new exceptions_1.WalletException(new Error('Target public key not found'));
            }
            if (!emailOTPId) {
                throw new exceptions_1.WalletException(new Error('Email OTP ID is required'));
            }
            if (!organizationId) {
                throw new exceptions_1.WalletException(new Error('Organization ID is required'));
            }
            const response = await client.post(otpVerifyPath, {
                targetPublicKey,
                otpId: emailOTPId,
                otpCode: args.otpCode,
                suborgID: organizationId,
                expirationSeconds: (expirationSeconds || consts_js_1.DEFAULT_TURNKEY_REFRESH_SECONDS)?.toString(),
            });
            return response?.data;
        }
        catch (e) {
            throw new exceptions_1.WalletException(new Error(e.message), {
                code: exceptions_1.UnspecifiedErrorCode,
                type: exceptions_1.ErrorType.WalletError,
                contextModule: 'turnkey-confirm-email-otp',
            });
        }
    }
}
exports.TurnkeyOtpWallet = TurnkeyOtpWallet;
