// @injectivelabs/wallet-core/dist/cjs/broadcaster/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletStrategyEmitterEventType = void 0;
var WalletStrategyEmitterEventType;
(function (WalletStrategyEmitterEventType) {
    WalletStrategyEmitterEventType["TransactionFail"] = "transaction-fail";
    WalletStrategyEmitterEventType["TransactionSigned"] = "transaction-signed";
    WalletStrategyEmitterEventType["TransactionSignStart"] = "transaction-sign-start";
    WalletStrategyEmitterEventType["TransactionBroadcastEnd"] = "transaction-broadcast-end";
    WalletStrategyEmitterEventType["WalletStrategyDisconnect"] = "wallet-strategy-disconnect";
    WalletStrategyEmitterEventType["TransactionBroadcastStart"] = "transaction-broadcast-start";
    WalletStrategyEmitterEventType["TransactionPreparationEnd"] = "transaction-preparation-end";
    WalletStrategyEmitterEventType["TransactionPreparationStart"] = "transaction-preparation-start";
})(WalletStrategyEmitterEventType || (exports.WalletStrategyEmitterEventType = WalletStrategyEmitterEventType = {}));
