// @keplr-wallet/proto-types/proto-types-gen/src/atomone/photon/v1/photon.ts
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal";

export const protobufPackage = "atomone.photon.v1";

/** Params defines the parameters for the x/photon module. */
export interface Params {
  /** Allow to mint photon or not */
  mintDisabled: boolean;
  /**
   * tx_fee_exceptions holds the msg type urls that are allowed to use some
   * different tx fee coins than photon.
   * A wildcard "*" can be used to allow all transactions to use any fee denom.
   */
  txFeeExceptions: string[];
}

function createBaseParams(): Params {
  return { mintDisabled: false, txFeeExceptions: [] };
}

export const Params = {
  encode(
    message: Params,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    if (message.mintDisabled === true) {
      writer.uint32(8).bool(message.mintDisabled);
    }
    for (const v of message.txFeeExceptions) {
      writer.uint32(18).string(v!);
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): Params {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseParams();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.mintDisabled = reader.bool();
          break;
        case 2:
          message.txFeeExceptions.push(reader.string());
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): Params {
    return {
      mintDisabled: isSet(object.mintDisabled)
        ? Boolean(object.mintDisabled)
        : false,
      txFeeExceptions: Array.isArray(object?.txFeeExceptions)
        ? object.txFeeExceptions.map((e: any) => String(e))
        : [],
    };
  },

  toJSON(message: Params): unknown {
    const obj: any = {};
    message.mintDisabled !== undefined &&
      (obj.mintDisabled = message.mintDisabled);
    if (message.txFeeExceptions) {
      obj.txFeeExceptions = message.txFeeExceptions.map((e) => e);
    } else {
      obj.txFeeExceptions = [];
    }
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<Params>, I>>(object: I): Params {
    const message = createBaseParams();
    message.mintDisabled = object.mintDisabled ?? false;
    message.txFeeExceptions = object.txFeeExceptions?.map((e) => e) || [];
    return message;
  },
};

type Builtin =
  | Date
  | Function
  | Uint8Array
  | string
  | number
  | boolean
  | undefined;

export type DeepPartial<T> = T extends Builtin
  ? T
  : T extends Array<infer U>
  ? Array<DeepPartial<U>>
  : T extends ReadonlyArray<infer U>
  ? ReadonlyArray<DeepPartial<U>>
  : T extends {}
  ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin
  ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & Record<
        Exclude<keyof I, KeysOfUnion<P>>,
        never
      >;

if (_m0.util.Long !== Long) {
  _m0.util.Long = Long as any;
  _m0.configure();
}

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
