// @injectivelabs/wallet-ts/dist/cjs/strategies/wallet-strategy/strategies/Leap.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable class-methods-use-this */
const ts_types_1 = require("@injectivelabs/ts-types");
const exceptions_1 = require("@injectivelabs/exceptions");
const sdk_ts_1 = require("@injectivelabs/sdk-ts");
const index_js_1 = require("../../../utils/wallets/leap/index.js");
const Base_js_1 = __importDefault(require("./Base.js"));
const enums_js_1 = require("../../../types/enums.js");
const cosmos_js_1 = require("../../../utils/cosmos.js");
class Leap extends Base_js_1.default {
    constructor(args) {
        super(args);
        this.chainId = args.chainId || ts_types_1.CosmosChainId.Injective;
        this.leapWallet = new index_js_1.LeapWallet(args.chainId);
    }
    getWalletDeviceType() {
        return __awaiter(this, void 0, void 0, function* () {
            const leapWallet = this.getLeapWallet();
            const key = yield leapWallet.getKey();
            return key.isNanoLedger
                ? Promise.resolve(enums_js_1.WalletDeviceType.Hardware)
                : Promise.resolve(enums_js_1.WalletDeviceType.Browser);
        });
    }
    enable() {
        return __awaiter(this, void 0, void 0, function* () {
            const leapWallet = this.getLeapWallet();
            return yield leapWallet.checkChainIdSupport();
        });
    }
    disconnect() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.listeners[enums_js_1.WalletEventListener.AccountChange]) {
                window.removeEventListener('leap_keystorechange', this.listeners[enums_js_1.WalletEventListener.AccountChange]);
            }
            this.listeners = {};
        });
    }
    getAddresses() {
        return __awaiter(this, void 0, void 0, function* () {
            const leapWallet = this.getLeapWallet();
            try {
                const accounts = yield leapWallet.getAccounts();
                return accounts.map((account) => account.address);
            }
            catch (e) {
                throw new exceptions_1.CosmosWalletException(new Error(e.message), {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: enums_js_1.WalletAction.GetAccounts,
                });
            }
        });
    }
    getSessionOrConfirm(address) {
        return __awaiter(this, void 0, void 0, function* () {
            return Promise.resolve(`0x${Buffer.from(`Confirmation for ${address} at time: ${Date.now()}`).toString('hex')}`);
        });
    }
    // eslint-disable-next-line class-methods-use-this
    sendEthereumTransaction(_transaction, _options) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new exceptions_1.CosmosWalletException(new Error('sendEthereumTransaction is not supported. Leap only supports sending cosmos transactions'), {
                code: exceptions_1.UnspecifiedErrorCode,
                context: enums_js_1.WalletAction.SendEthereumTransaction,
            });
        });
    }
    sendTransaction(transaction, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { leapWallet } = this;
            const txRaw = (0, sdk_ts_1.createTxRawFromSigResponse)(transaction);
            if (!options.endpoints) {
                throw new exceptions_1.CosmosWalletException(new Error('You have to pass endpoints within the options to broadcast transaction'));
            }
            try {
                const txHash = yield leapWallet.broadcastTx(txRaw);
                return yield (0, sdk_ts_1.waitTxBroadcasted)(txHash, options);
            }
            catch (e) {
                if (e instanceof exceptions_1.TransactionException) {
                    throw e;
                }
                throw new exceptions_1.TransactionException(new Error(e.message), {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: enums_js_1.WalletAction.SendTransaction,
                });
            }
        });
    }
    /** @deprecated */
    signTransaction(transaction, injectiveAddress) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.signCosmosTransaction(Object.assign(Object.assign({}, transaction), { address: injectiveAddress }));
        });
    }
    signAminoCosmosTransaction(_transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new exceptions_1.CosmosWalletException(new Error('This wallet does not support signing using amino'), {
                code: exceptions_1.UnspecifiedErrorCode,
                context: enums_js_1.WalletAction.SendTransaction,
            });
        });
    }
    signCosmosTransaction(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            const leapWallet = this.getLeapWallet();
            const signer = yield leapWallet.getOfflineSigner();
            const signDoc = (0, sdk_ts_1.createSignDocFromTransaction)(transaction);
            try {
                return yield signer.signDirect(transaction.address, (0, cosmos_js_1.createCosmosSignDocFromSignDoc)(signDoc));
            }
            catch (e) {
                throw new exceptions_1.CosmosWalletException(new Error(e.message), {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: enums_js_1.WalletAction.SendTransaction,
                });
            }
        });
    }
    signArbitrary(signer, data) {
        return __awaiter(this, void 0, void 0, function* () {
            const leapWallet = this.getLeapWallet();
            const leap = yield leapWallet.getLeapWallet();
            try {
                const signature = yield leap.signArbitrary(this.chainId, signer, data);
                return signature.signature;
            }
            catch (e) {
                throw new exceptions_1.CosmosWalletException(new Error(e.message), {
                    code: exceptions_1.UnspecifiedErrorCode,
                    context: enums_js_1.WalletAction.SignArbitrary,
                });
            }
        });
    }
    signEip712TypedData(_eip712TypedData, _address) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new exceptions_1.CosmosWalletException(new Error('This wallet does not support signing Ethereum transactions'), {
                code: exceptions_1.UnspecifiedErrorCode,
                context: enums_js_1.WalletAction.SendTransaction,
            });
        });
    }
    getEthereumChainId() {
        return __awaiter(this, void 0, void 0, function* () {
            throw new exceptions_1.CosmosWalletException(new Error('getEthereumChainId is not supported on Leap'), {
                code: exceptions_1.UnspecifiedErrorCode,
                context: enums_js_1.WalletAction.GetChainId,
            });
        });
    }
    getEthereumTransactionReceipt(_txHash) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new exceptions_1.CosmosWalletException(new Error('getEthereumTransactionReceipt is not supported on Leap'), {
                code: exceptions_1.UnspecifiedErrorCode,
                context: enums_js_1.WalletAction.GetEthereumTransactionReceipt,
            });
        });
    }
    getPubKey() {
        return __awaiter(this, void 0, void 0, function* () {
            const keplrWallet = this.getLeapWallet();
            const key = yield keplrWallet.getKey();
            return Buffer.from(key.pubKey).toString('base64');
        });
    }
    onAccountChange(callback) {
        return __awaiter(this, void 0, void 0, function* () {
            const listener = () => __awaiter(this, void 0, void 0, function* () {
                const [account] = yield this.getAddresses();
                callback(account);
            });
            this.listeners = {
                [enums_js_1.WalletEventListener.AccountChange]: listener,
            };
            window.addEventListener('leap_keystorechange', listener);
        });
    }
    getLeapWallet() {
        const { leapWallet } = this;
        if (!leapWallet) {
            throw new exceptions_1.CosmosWalletException(new Error('Please install the Leap wallet extension'), {
                code: exceptions_1.UnspecifiedErrorCode,
                type: exceptions_1.ErrorType.WalletNotInstalledError,
                context: enums_js_1.WalletAction.SignTransaction,
            });
        }
        return leapWallet;
    }
}
exports.default = Leap;
