// @injectivelabs/ts-types/dist/cjs/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EIP712Version = exports.MsgStatus = exports.MsgType = exports.ChainId = exports.EthereumChainId = void 0;
var EthereumChainId;
(function (EthereumChainId) {
    EthereumChainId[EthereumChainId["Mainnet"] = 1] = "Mainnet";
    EthereumChainId[EthereumChainId["Ropsten"] = 3] = "Ropsten";
    EthereumChainId[EthereumChainId["Rinkeby"] = 4] = "Rinkeby";
    EthereumChainId[EthereumChainId["Goerli"] = 5] = "Goerli";
    EthereumChainId[EthereumChainId["Kovan"] = 42] = "Kovan";
    EthereumChainId[EthereumChainId["Injective"] = 888] = "Injective";
    EthereumChainId[EthereumChainId["Sepolia"] = 11155111] = "Sepolia";
    EthereumChainId[EthereumChainId["Ganache"] = 1337] = "Ganache";
    EthereumChainId[EthereumChainId["HardHat"] = 31337] = "HardHat";
    EthereumChainId[EthereumChainId["DevnetEvm"] = 1778] = "DevnetEvm";
    EthereumChainId[EthereumChainId["TestnetEvm"] = 1439] = "TestnetEvm";
})(EthereumChainId || (exports.EthereumChainId = EthereumChainId = {}));
var ChainId;
(function (ChainId) {
    ChainId["Mainnet"] = "injective-1";
    ChainId["Testnet"] = "injective-888";
    ChainId["Devnet"] = "injective-777";
})(ChainId || (exports.ChainId = ChainId = {}));
var MsgType;
(function (MsgType) {
    // Authz
    MsgType["MsgExec"] = "cosmos.authz.v1beta1.MsgExec";
    MsgType["MsgGrant"] = "cosmos.authz.v1beta1.MsgGrant";
    MsgType["MsgRevoke"] = "cosmos.authz.v1beta1.MsgRevoke";
    // Bank
    MsgType["MsgSend"] = "cosmos.bank.v1beta1.MsgSend";
    MsgType["MsgMultiSend"] = "cosmos.bank.v1beta1.MsgMultiSend";
    // Distribution
    MsgType["MsgSetWithdrawAddress"] = "cosmos.distribution.v1beta1.MsgSetWithdrawAddress";
    MsgType["MsgWithdrawDelegatorReward"] = "cosmos.distribution.v1beta1.MsgWithdrawDelegatorReward";
    MsgType["MsgWithdrawValidatorCommission"] = "cosmos.distribution.v1beta1.MsgWithdrawValidatorCommission";
    // Feegrant
    MsgType["MsgGrantAllowance"] = "cosmos.feegrant.v1beta1.MsgGrantAllowance";
    MsgType["MsgRevokeAllowance"] = "cosmos.feegrant.v1beta1.MsgRevokeAllowance";
    // Governance
    MsgType["MsgVote"] = "cosmos.gov.v1beta1.MsgVote";
    MsgType["MsgDepositCosmos"] = "cosmos.gov.v1beta1.MsgDeposit";
    MsgType["MsgSubmitProposal"] = "cosmos.gov.v1beta1.MsgSubmitProposal";
    // Slashing
    MsgType["MsgUnjail"] = "cosmos.slashing.v1beta1.MsgUnjail";
    // Staking
    MsgType["MsgDelegate"] = "cosmos.staking.v1beta1.MsgDelegate";
    MsgType["MsgUndelegate"] = "cosmos.staking.v1beta1.MsgUndelegate";
    MsgType["MsgEditValidator"] = "cosmos.staking.v1beta1.MsgEditValidator";
    MsgType["MsgBeginRedelegate"] = "cosmos.staking.v1beta1.MsgBeginRedelegate";
    MsgType["MsgCreateValidator"] = "cosmos.staking.v1beta1.MsgCreateValidator";
    MsgType["MsgCancelUnbondingDelegation"] = "cosmos.staking.v1beta1.MsgCancelUnbondingDelegation";
    MsgType["MsgStoreCode"] = "cosmwasm.wasm.v1.MsgStoreCode";
    MsgType["MsgUpdateAdmin"] = "cosmwasm.wasm.v1.MsgUpdateAdmin";
    MsgType["MsgExecuteContract"] = "cosmwasm.wasm.v1.MsgExecuteContract";
    MsgType["MsgMigrateContract"] = "cosmwasm.wasm.v1.MsgMigrateContract";
    MsgType["MsgInstantiateContract"] = "cosmwasm.wasm.v1.MsgInstantiateContract";
    MsgType["MsgInstantiateContract2"] = "cosmwasm.wasm.v1.MsgInstantiateContract2";
    // Transfer
    MsgType["MsgTransfer"] = "ibc.applications.transfer.v1.MsgTransfer";
    // Channel
    MsgType["MsgTimeout"] = "ibc.core.channel.v1.MsgTimeout";
    MsgType["MsgRecvPacket"] = "ibc.core.channel.v1.MsgRecvPacket";
    MsgType["MsgChannelOpenAck"] = "ibc.core.channel.v1.MsgChannelOpenAck";
    MsgType["MsgChannelOpenTry"] = "ibc.core.channel.v1.MsgChannelOpenTry";
    MsgType["MsgAcknowledgement"] = "ibc.core.channel.v1.MsgAcknowledgement";
    MsgType["MsgChannelOpenInit"] = "ibc.core.channel.v1.MsgChannelOpenInit";
    MsgType["MsgChannelOpenConfirm"] = "ibc.core.channel.v1.MsgChannelOpenConfirm";
    // Client
    MsgType["MsgCreateClient"] = "ibc.core.client.v1.MsgCreateClient";
    MsgType["MsgUpdateClient"] = "ibc.core.client.v1.MsgUpdateClient";
    // Connection
    MsgType["MsgConnectionOpenAck"] = "ibc.core.connection.v1.MsgConnectionOpenAck";
    MsgType["MsgConnectionOpenTry"] = "ibc.core.connection.v1.MsgConnectionOpenTry";
    MsgType["MsgConnectionOpenInit"] = "ibc.core.connection.v1.MsgConnectionOpenInit";
    MsgType["MsgConnectionOpenConfirm"] = "ibc.core.connection.v1.MsgConnectionOpenConfirm";
    // Auction
    MsgType["MsgBid"] = "injective.auction.v1beta1.MsgBid";
    MsgType["MsgAuctionUpdateParams"] = "injective.auction.v1beta1.MsgUpdateParams";
    // ERC20
    MsgType["MsgErc20UpdateParams"] = "injective.erc20.v1beta1.MsgUpdateParams";
    MsgType["MsgCreateTokenPair"] = "injective.erc20.v1beta1.MsgCreateTokenPair";
    MsgType["MsgDeleteTokenPair"] = "injective.erc20.v1beta1.MsgDeleteTokenPair";
    // EVM
    MsgType["LegacyTx"] = "injective.evm.v1.LegacyTx";
    MsgType["AccessListTx"] = "injective.evm.v1.AccessListTx";
    MsgType["DynamicFeeTx"] = "injective.evm.v1.DynamicFeeTx";
    MsgType["MsgEthereumTx"] = "injective.evm.v1.MsgEthereumTx";
    MsgType["MsgEvmUpdateParams"] = "injective.evm.v1.MsgUpdateParams";
    MsgType["ExtensionOptionsEthereumTx"] = "injective.evm.v1.ExtensionOptionsEthereumTx";
    // Exchange
    MsgType["OrderData"] = "injective.exchange.v1beta1.OrderData";
    MsgType["MsgDeposit"] = "injective.exchange.v1beta1.MsgDeposit";
    MsgType["MsgSignDoc"] = "injective.exchange.v1beta1.MsgSignDoc";
    MsgType["MsgSignData"] = "injective.exchange.v1beta1.MsgSignData";
    MsgType["MsgWithdraw"] = "injective.exchange.v1beta1.MsgWithdraw";
    MsgType["MsgRewardsOptOut"] = "injective.exchange.v1beta1.MsgRewardsOptOut";
    MsgType["MsgCancelSpotOrder"] = "injective.exchange.v1beta1.MsgCancelSpotOrder";
    MsgType["MsgExchangeUpdateParams"] = "injective.exchange.v1beta1.MsgUpdateParams";
    MsgType["MsgExternalTransfer"] = "injective.exchange.v1beta1.MsgExternalTransfer";
    MsgType["MsgUpdateSpotMarket"] = "injective.exchange.v1beta1.MsgUpdateSpotMarket";
    MsgType["MsgBatchUpdateOrders"] = "injective.exchange.v1beta1.MsgBatchUpdateOrders";
    MsgType["MsgLiquidatePosition"] = "injective.exchange.v1beta1.MsgLiquidatePosition";
    MsgType["MsgActivateStakeGrant"] = "injective.exchange.v1beta1.MsgActivateStakeGrant";
    MsgType["MsgReclaimLockedFunds"] = "injective.exchange.v1beta1.MsgReclaimLockedFunds";
    MsgType["MsgSubaccountTransfer"] = "injective.exchange.v1beta1.MsgSubaccountTransfer";
    MsgType["SpotMarketOrderResults"] = "injective.exchange.v1beta1.SpotMarketOrderResults";
    MsgType["MsgAuthorizeStakeGrants"] = "injective.exchange.v1beta1.MsgAuthorizeStakeGrants";
    MsgType["MsgCreateSpotLimitOrder"] = "injective.exchange.v1beta1.MsgCreateSpotLimitOrder";
    MsgType["MsgBatchCancelSpotOrders"] = "injective.exchange.v1beta1.MsgBatchCancelSpotOrders";
    MsgType["MsgCancelDerivativeOrder"] = "injective.exchange.v1beta1.MsgCancelDerivativeOrder";
    MsgType["MsgCreateSpotMarketOrder"] = "injective.exchange.v1beta1.MsgCreateSpotMarketOrder";
    MsgType["MsgEmergencySettleMarket"] = "injective.exchange.v1beta1.MsgEmergencySettleMarket";
    MsgType["MsgDecreasePositionMargin"] = "injective.exchange.v1beta1.MsgDecreasePositionMargin";
    MsgType["MsgIncreasePositionMargin"] = "injective.exchange.v1beta1.MsgIncreasePositionMargin";
    MsgType["MsgUpdateDerivativeMarket"] = "injective.exchange.v1beta1.MsgUpdateDerivativeMarket";
    MsgType["MsgInstantSpotMarketLaunch"] = "injective.exchange.v1beta1.MsgInstantSpotMarketLaunch";
    MsgType["MsgCancelBinaryOptionsOrder"] = "injective.exchange.v1beta1.MsgCancelBinaryOptionsOrder";
    MsgType["DerivativeMarketOrderResults"] = "injective.exchange.v1beta1.DerivativeMarketOrderResults";
    MsgType["MsgBatchExchangeModification"] = "injective.exchange.v1beta1.MsgBatchExchangeModification";
    MsgType["MsgPrivilegedExecuteContract"] = "injective.exchange.v1beta1.MsgPrivilegedExecuteContract";
    MsgType["MsgBatchCreateSpotLimitOrders"] = "injective.exchange.v1beta1.MsgBatchCreateSpotLimitOrders";
    MsgType["MsgCreateDerivativeLimitOrder"] = "injective.exchange.v1beta1.MsgCreateDerivativeLimitOrder";
    MsgType["MsgBatchCancelDerivativeOrders"] = "injective.exchange.v1beta1.MsgBatchCancelDerivativeOrders";
    MsgType["MsgCreateDerivativeMarketOrder"] = "injective.exchange.v1beta1.MsgCreateDerivativeMarketOrder";
    MsgType["MsgInstantPerpetualMarketLaunch"] = "injective.exchange.v1beta1.MsgInstantPerpetualMarketLaunch";
    MsgType["MsgCreateBinaryOptionsLimitOrder"] = "injective.exchange.v1beta1.MsgCreateBinaryOptionsLimitOrder";
    MsgType["MsgAdminUpdateBinaryOptionsMarket"] = "injective.exchange.v1beta1.MsgAdminUpdateBinaryOptionsMarket";
    MsgType["MsgBatchCancelBinaryOptionsOrders"] = "injective.exchange.v1beta1.MsgBatchCancelBinaryOptionsOrders";
    MsgType["MsgCreateBinaryOptionsMarketOrder"] = "injective.exchange.v1beta1.MsgCreateBinaryOptionsMarketOrder";
    MsgType["MsgInstantExpiryFuturesMarketLaunch"] = "injective.exchange.v1beta1.MsgInstantExpiryFuturesMarketLaunch";
    MsgType["MsgBatchCreateDerivativeLimitOrders"] = "injective.exchange.v1beta1.MsgBatchCreateDerivativeLimitOrders";
    MsgType["MsgInstantBinaryOptionsMarketLaunch"] = "injective.exchange.v1beta1.MsgInstantBinaryOptionsMarketLaunch";
    // Exchange V2
    MsgType["OrderDataV2"] = "injective.exchange.v2.OrderData";
    MsgType["MsgDepositV2"] = "injective.exchange.v2.MsgDeposit";
    MsgType["MsgSignDocV2"] = "injective.exchange.v2.MsgSignDoc";
    MsgType["MsgSignDataV2"] = "injective.exchange.v2.MsgSignData";
    MsgType["MsgWithdrawV2"] = "injective.exchange.v2.MsgWithdraw";
    MsgType["MsgFeeDiscountV2"] = "injective.exchange.v2.MsgFeeDiscount";
    MsgType["MsgUpdateParamsV2"] = "injective.exchange.v2.MsgUpdateParams";
    MsgType["MsgRewardsOptOutV2"] = "injective.exchange.v2.MsgRewardsOptOut";
    MsgType["MsgExchangeEnableV2"] = "injective.exchange.v2.MsgExchangeEnable";
    MsgType["MsgCancelSpotOrderV2"] = "injective.exchange.v2.MsgCancelSpotOrder";
    MsgType["MsgExternalTransferV2"] = "injective.exchange.v2.MsgExternalTransfer";
    MsgType["MsgSpotMarketLaunchV2"] = "injective.exchange.v2.MsgSpotMarketLaunch";
    MsgType["MsgUpdateSpotMarketV2"] = "injective.exchange.v2.MsgUpdateSpotMarket";
    MsgType["MsgBatchUpdateOrdersV2"] = "injective.exchange.v2.MsgBatchUpdateOrders";
    MsgType["MsgLiquidatePositionV2"] = "injective.exchange.v2.MsgLiquidatePosition";
    MsgType["MsgActivateStakeGrantV2"] = "injective.exchange.v2.MsgActivateStakeGrant";
    MsgType["MsgReclaimLockedFundsV2"] = "injective.exchange.v2.MsgReclaimLockedFunds";
    MsgType["MsgSubaccountTransferV2"] = "injective.exchange.v2.MsgSubaccountTransfer";
    MsgType["SpotMarketOrderResultsV2"] = "injective.exchange.v2.SpotMarketOrderResults";
    MsgType["MsgAuthorizeStakeGrantsV2"] = "injective.exchange.v2.MsgAuthorizeStakeGrants";
    MsgType["MsgCreateSpotLimitOrderV2"] = "injective.exchange.v2.MsgCreateSpotLimitOrder";
    MsgType["MsgBatchCancelSpotOrdersV2"] = "injective.exchange.v2.MsgBatchCancelSpotOrders";
    MsgType["MsgCancelDerivativeOrderV2"] = "injective.exchange.v2.MsgCancelDerivativeOrder";
    MsgType["MsgCreateSpotMarketOrderV2"] = "injective.exchange.v2.MsgCreateSpotMarketOrder";
    MsgType["MsgEmergencySettleMarketV2"] = "injective.exchange.v2.MsgEmergencySettleMarket";
    MsgType["MsgPerpetualMarketLaunchV2"] = "injective.exchange.v2.MsgPerpetualMarketLaunch";
    MsgType["MsgSpotMarketParamUpdateV2"] = "injective.exchange.v2.MsgSpotMarketParamUpdate";
    MsgType["MsgDecreasePositionMarginV2"] = "injective.exchange.v2.MsgDecreasePositionMargin";
    MsgType["MsgIncreasePositionMarginV2"] = "injective.exchange.v2.MsgIncreasePositionMargin";
    MsgType["MsgMarketForcedSettlementV2"] = "injective.exchange.v2.MsgMarketForcedSettlement";
    MsgType["MsgUpdateDerivativeMarketV2"] = "injective.exchange.v2.MsgUpdateDerivativeMarket";
    MsgType["MsgBatchCommunityPoolSpendV2"] = "injective.exchange.v2.MsgBatchCommunityPoolSpend";
    MsgType["MsgInstantSpotMarketLaunchV2"] = "injective.exchange.v2.MsgInstantSpotMarketLaunch";
    MsgType["MsgCancelBinaryOptionsOrderV2"] = "injective.exchange.v2.MsgCancelBinaryOptionsOrder";
    MsgType["DerivativeMarketOrderResultsV2"] = "injective.exchange.v2.DerivativeMarketOrderResults";
    MsgType["MsgBatchExchangeModificationV2"] = "injective.exchange.v2.MsgBatchExchangeModification";
    MsgType["MsgBinaryOptionsMarketLaunchV2"] = "injective.exchange.v2.MsgBinaryOptionsMarketLaunch";
    MsgType["MsgExpiryFuturesMarketLaunchV2"] = "injective.exchange.v2.MsgExpiryFuturesMarketLaunch";
    MsgType["MsgPrivilegedExecuteContractV2"] = "injective.exchange.v2.MsgPrivilegedExecuteContract";
    MsgType["MsgBatchCreateSpotLimitOrdersV2"] = "injective.exchange.v2.MsgBatchCreateSpotLimitOrders";
    MsgType["MsgCreateDerivativeLimitOrderV2"] = "injective.exchange.v2.MsgCreateDerivativeLimitOrder";
    MsgType["MsgBatchCancelDerivativeOrdersV2"] = "injective.exchange.v2.MsgBatchCancelDerivativeOrders";
    MsgType["MsgCreateDerivativeMarketOrderV2"] = "injective.exchange.v2.MsgCreateDerivativeMarketOrder";
    MsgType["MsgDerivativeMarketParamUpdateV2"] = "injective.exchange.v2.MsgDerivativeMarketParamUpdate";
    MsgType["MsgTradingRewardCampaignLaunchV2"] = "injective.exchange.v2.MsgTradingRewardCampaignLaunch";
    MsgType["MsgTradingRewardCampaignUpdateV2"] = "injective.exchange.v2.MsgTradingRewardCampaignUpdate";
    MsgType["MsgInstantPerpetualMarketLaunchV2"] = "injective.exchange.v2.MsgInstantPerpetualMarketLaunch";
    MsgType["MsgCreateBinaryOptionsLimitOrderV2"] = "injective.exchange.v2.MsgCreateBinaryOptionsLimitOrder";
    MsgType["MsgAdminUpdateBinaryOptionsMarketV2"] = "injective.exchange.v2.MsgAdminUpdateBinaryOptionsMarket";
    MsgType["MsgBatchCancelBinaryOptionsOrdersV2"] = "injective.exchange.v2.MsgBatchCancelBinaryOptionsOrders";
    MsgType["MsgBinaryOptionsMarketParamUpdateV2"] = "injective.exchange.v2.MsgBinaryOptionsMarketParamUpdate";
    MsgType["MsgCreateBinaryOptionsMarketOrderV2"] = "injective.exchange.v2.MsgCreateBinaryOptionsMarketOrder";
    MsgType["MsgBatchCreateDerivativeLimitOrdersV2"] = "injective.exchange.v2.MsgBatchCreateDerivativeLimitOrders";
    MsgType["MsgInstantBinaryOptionsMarketLaunchV2"] = "injective.exchange.v2.MsgInstantBinaryOptionsMarketLaunch";
    MsgType["MsgInstantExpiryFuturesMarketLaunchV2"] = "injective.exchange.v2.MsgInstantExpiryFuturesMarketLaunch";
    MsgType["MsgTradingRewardPendingPointsUpdateV2"] = "injective.exchange.v2.MsgTradingRewardPendingPointsUpdate";
    MsgType["MsgAtomicMarketOrderFeeMultiplierScheduleV2"] = "injective.exchange.v2.MsgAtomicMarketOrderFeeMultiplierSchedule";
    // Insurance
    MsgType["MsgUnderwrite"] = "injective.insurance.v1beta1.MsgUnderwrite";
    MsgType["MsgInsuranceUpdateParams"] = "injective.insurance.v1beta1.MsgUpdateParams";
    MsgType["MsgRequestRedemption"] = "injective.insurance.v1beta1.MsgRequestRedemption";
    MsgType["MsgCreateInsuranceFund"] = "injective.insurance.v1beta1.MsgCreateInsuranceFund";
    // Oracle
    MsgType["MsgRelayBandRates"] = "injective.oracle.v1beta1.MsgRelayBandRates";
    MsgType["MsgRelayPythPrices"] = "injective.oracle.v1beta1.MsgRelayPythPrices";
    MsgType["MsgOracleUpdateParams"] = "injective.oracle.v1beta1.MsgUpdateParams";
    MsgType["MsgRelayStorkPrices"] = "injective.oracle.v1beta1.MsgRelayStorkPrices";
    MsgType["MsgRelayProviderPrices"] = "injective.oracle.v1beta1.MsgRelayProviderPrices";
    MsgType["MsgRelayPriceFeedPrice"] = "injective.oracle.v1beta1.MsgRelayPriceFeedPrice";
    MsgType["MsgRequestBandIBCRates"] = "injective.oracle.v1beta1.MsgRequestBandIBCRates";
    MsgType["MsgRelayCoinbaseMessages"] = "injective.oracle.v1beta1.MsgRelayCoinbaseMessages";
    // OCR
    MsgType["MsgTransmit"] = "injective.ocr.v1beta1.MsgTransmit";
    MsgType["MsgSetPayees"] = "injective.ocr.v1beta1.MsgSetPayees";
    MsgType["MsgCreateFeed"] = "injective.ocr.v1beta1.MsgCreateFeed";
    MsgType["MsgUpdateFeed"] = "injective.ocr.v1beta1.MsgUpdateFeed";
    MsgType["MsgOcrUpdateParams"] = "injective.ocr.v1beta1.MsgUpdateParams";
    MsgType["MsgAcceptPayeeship"] = "injective.ocr.v1beta1.MsgAcceptPayeeship";
    MsgType["MsgTransferPayeeship"] = "injective.ocr.v1beta1.MsgTransferPayeeship";
    MsgType["MsgFundFeedRewardPool"] = "injective.ocr.v1beta1.MsgFundFeedRewardPool";
    MsgType["MsgWithdrawFeedRewardPool"] = "injective.ocr.v1beta1.MsgWithdrawFeedRewardPool";
    // Peggy (Bridge)
    MsgType["MsgSendToEth"] = "injective.peggy.v1.MsgSendToEth";
    MsgType["MsgConfirmBatch"] = "injective.peggy.v1.MsgConfirmBatch";
    MsgType["MsgDepositClaim"] = "injective.peggy.v1.MsgDepositClaim";
    MsgType["MsgRequestBatch"] = "injective.peggy.v1.MsgRequestBatch";
    MsgType["MsgValsetConfirm"] = "injective.peggy.v1.MsgValsetConfirm";
    MsgType["MsgWithdrawClaim"] = "injective.peggy.v1.MsgWithdrawClaim";
    MsgType["MsgERC20DeployedClaim"] = "injective.peggy.v1.MsgERC20DeployedClaim";
    MsgType["MsgValsetUpdatedClaim"] = "injective.peggy.v1.MsgValsetUpdatedClaim";
    MsgType["MsgRevokeEthereumBlacklist"] = "injective.peggy.v1.MsgRevokeEthereumBlacklist";
    MsgType["MsgSetOrchestratorAddresses"] = "injective.peggy.v1.MsgSetOrchestratorAddresses";
    MsgType["MsgBlacklistEthereumAddresses"] = "injective.peggy.v1.MsgBlacklistEthereumAddresses";
    // Permissions
    MsgType["MsgClaimVoucher"] = "injective.permissions.v1beta1.MsgClaimVoucher";
    MsgType["MsgCreateNamespace"] = "injective.permissions.v1beta1.MsgCreateNamespace";
    MsgType["MsgDeleteNamespace"] = "injective.permissions.v1beta1.MsgDeleteNamespace";
    MsgType["MsgUpdateNamespace"] = "injective.permissions.v1beta1.MsgUpdateNamespace";
    MsgType["MsgUpdateActorRoles"] = "injective.permissions.v1beta1.MsgUpdateActorRoles";
    MsgType["MsgPermissionsUpdateParams"] = "injective.permissions.v1beta1.MsgUpdateParams";
    MsgType["MsgUpdateNamespaceRoles"] = "injective.permissions.v1beta1.MsgUpdateNamespaceRoles";
    MsgType["MsgRevokeNamespaceRoles"] = "injective.permissions.v1beta1.MsgRevokeNamespaceRoles";
    MsgType["MsgSetWasmHook"] = "injective.permissions.v1beta1.MsgUpdateNamespace.MsgSetWasmHook";
    MsgType["MsgSetMintsPaused"] = "injective.permissions.v1beta1.MsgUpdateNamespace.MsgSetMintsPaused";
    MsgType["MsgSetSendsPaused"] = "injective.permissions.v1beta1.MsgUpdateNamespace.MsgSetSendsPaused";
    MsgType["MsgSetBurnsPaused"] = "injective.permissions.v1beta1.MsgUpdateNamespace.MsgSetBurnsPaused";
    MsgType["MsgUpdateNamespaceSetContractHook"] = "injective.permissions.v1beta1.MsgUpdateNamespace.SetContractHook";
    // TokenFactory
    MsgType["MsgBurn"] = "injective.tokenfactory.v1beta1.MsgBurn";
    MsgType["MsgMint"] = "injective.tokenfactory.v1beta1.MsgMint";
    MsgType["MsgCreateDenom"] = "injective.tokenfactory.v1beta1.MsgCreateDenom";
    MsgType["MsgChangeAdmin"] = "injective.tokenfactory.v1beta1.MsgChangeAdmin";
    MsgType["MsgSetDenomMetadata"] = "injective.tokenfactory.v1beta1.MsgSetDenomMetadata";
    MsgType["MsgTokenFactoryUpdateParams"] = "injective.tokenfactory.v1beta1.MsgUpdateParams";
    MsgType["MsgSetDenomMetadataAdminBurnDisabled"] = "injective.tokenfactory.v1beta1.MsgSetDenomMetadata.AdminBurnDisabled";
    // Wasmx
    MsgType["MsgUpdateContract"] = "injective.wasmx.v1.MsgUpdateContract";
    MsgType["MsgWasmxUpdateParams"] = "injective.wasmx.v1.MsgUpdateParams";
    MsgType["MsgActivateContract"] = "injective.wasmx.v1.MsgActivateContract";
    MsgType["MsgRegisterContract"] = "injective.wasmx.v1.MsgRegisterContract";
    MsgType["MsgDeactivateContract"] = "injective.wasmx.v1.MsgDeactivateContract";
    MsgType["MsgExecuteContractCompat"] = "injective.wasmx.v1.MsgExecuteContractCompat";
    // TxFees
    MsgType["MsgTxFeesUpdateParams"] = "injective.txfees.v1beta1.MsgUpdateParams";
})(MsgType || (exports.MsgType = MsgType = {}));
var MsgStatus;
(function (MsgStatus) {
    MsgStatus["Success"] = "success";
    MsgStatus["Fail"] = "fail";
})(MsgStatus || (exports.MsgStatus = MsgStatus = {}));
var EIP712Version;
(function (EIP712Version) {
    EIP712Version["V1"] = "v1";
    EIP712Version["V2"] = "v2";
})(EIP712Version || (exports.EIP712Version = EIP712Version = {}));
