// @injectivelabs/wallet-ledger/dist/cjs/strategy/Ledger/hw/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ledgerhq_hw_transport_webhid_1 = require("@bangjelkoski/ledgerhq-hw-transport-webhid");
const ledgerhq_hw_transport_webusb_1 = require("@bangjelkoski/ledgerhq-hw-transport-webusb");
const ledgerhq_hw_app_eth_1 = require("@bangjelkoski/ledgerhq-hw-app-eth");
const exceptions_1 = require("@injectivelabs/exceptions");
const AccountManager_js_1 = __importDefault(require("./AccountManager.js"));
class LedgerTransport {
    ledger = null;
    accountManager = null;
    static async getTransport() {
        try {
            if (await ledgerhq_hw_transport_webhid_1.TransportWebHID.isSupported()) {
                const list = await ledgerhq_hw_transport_webhid_1.TransportWebHID.list();
                if (list.length > 0 && list[0].opened) {
                    return new ledgerhq_hw_transport_webhid_1.TransportWebHID(list[0]);
                }
                const existing = await ledgerhq_hw_transport_webhid_1.TransportWebHID.openConnected().catch(() => null);
                if (existing) {
                    return existing;
                }
                return await ledgerhq_hw_transport_webhid_1.TransportWebHID.request();
            }
            if (await ledgerhq_hw_transport_webusb_1.TransportWebUSB.isSupported()) {
                const existing = await ledgerhq_hw_transport_webusb_1.TransportWebUSB.openConnected().catch(() => null);
                if (existing) {
                    return existing;
                }
                return await ledgerhq_hw_transport_webusb_1.TransportWebUSB.request();
            }
        }
        catch (e) {
            throw new exceptions_1.LedgerException(new Error(e.message));
        }
        return ledgerhq_hw_transport_webusb_1.TransportWebUSB.request();
    }
    async getInstance() {
        if (!this.ledger) {
            const transport = await LedgerTransport.getTransport();
            this.ledger = new ledgerhq_hw_app_eth_1.Eth(transport);
            transport.on('disconnect', () => {
                this.ledger = null;
                this.accountManager = null;
            });
        }
        return this.ledger;
    }
    async getAccountManager() {
        if (!this.accountManager) {
            this.accountManager = new AccountManager_js_1.default(await this.getInstance());
        }
        return this.accountManager;
    }
    async refresh() {
        if (!this.ledger) {
            return new LedgerTransport();
        }
        this.ledger.transport.close();
        return new LedgerTransport();
    }
}
exports.default = LedgerTransport;
