// @injectivelabs/wallet-ts/dist/cjs/strategies/wallet-strategy/strategies/LedgerCosmos/hw/AccountManager.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const constants_js_1 = require("../../../constants.js");
class AccountManager {
    constructor(ledger) {
        this.wallets = [];
        this.getLedgerDerivationPathBasedOnType = ({ fullBaseDerivationPath, index, }) => {
            return `${fullBaseDerivationPath}/${index}'/0/0`;
        };
        this.ledger = ledger;
        this.wallets = [];
    }
    getWallets(baseDerivationPath) {
        return __awaiter(this, void 0, void 0, function* () {
            const { start, end } = this.getOffset();
            /**
             * 1. Wallets are not yet fetched at all,
             * 2. Wallets are not yet fetched for that offset
             */
            if (!this.hasWallets() || !this.hasWalletsInOffset(start)) {
                yield this.getWalletsBasedOnIndex({
                    start,
                    end,
                    baseDerivationPath,
                });
            }
            return this.wallets.slice(start, end);
        });
    }
    getWalletsBasedOnIndex(_a) {
        return __awaiter(this, arguments, void 0, function* ({ start, end, baseDerivationPath, }) {
            for (let index = start; index < end; index += 1) {
                const path = this.getLedgerDerivationPathBasedOnType({
                    fullBaseDerivationPath: baseDerivationPath,
                    index,
                });
                const { address, publicKey } = yield this.ledger.getAddress(path, 'inj');
                this.wallets.push({
                    publicKey,
                    baseDerivationPath,
                    address: address.toLowerCase(),
                    derivationPath: path,
                });
            }
        });
    }
    hasWallets() {
        return this.wallets.length > 0;
    }
    hasWalletsInOffset(offset) {
        return this.wallets.length > offset;
    }
    getOffset() {
        const totalWallets = this.wallets.length;
        const nextBatchStart = totalWallets;
        const nextBatchEnd = totalWallets + constants_js_1.DEFAULT_NUM_ADDRESSES_TO_FETCH;
        return {
            start: nextBatchStart,
            end: nextBatchEnd,
        };
    }
    hasWalletForAddress(address) {
        return (this.wallets.find((wallet) => wallet.address.toLowerCase() === address.toLowerCase()) !== undefined);
    }
    getWalletForAddress(address) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.wallets.find((wallet) => wallet.address.toLowerCase() === address.toLowerCase());
        });
    }
    reset() {
        this.wallets = [];
    }
}
exports.default = AccountManager;
