// @injectivelabs/wallet-ts/dist/cjs/strategies/wallet-strategy/strategies/LedgerCosmos/hw/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const hw_transport_webusb_1 = __importDefault(require("@ledgerhq/hw-transport-webusb"));
const hw_transport_webhid_1 = __importDefault(require("@ledgerhq/hw-transport-webhid"));
const hw_app_cosmos_1 = __importDefault(require("@ledgerhq/hw-app-cosmos"));
const exceptions_1 = require("@injectivelabs/exceptions");
const AccountManager_js_1 = __importDefault(require("./AccountManager.js"));
class LedgerTransport {
    constructor() {
        this.ledger = null;
        this.accountManager = null;
    }
    static getTransport() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (yield hw_transport_webhid_1.default.isSupported()) {
                    return yield hw_transport_webhid_1.default.create();
                }
                if (yield hw_transport_webusb_1.default.isSupported()) {
                    return yield hw_transport_webhid_1.default.create();
                }
            }
            catch (e) {
                throw new exceptions_1.LedgerCosmosException(new Error(e.message));
            }
            return yield hw_transport_webhid_1.default.create();
        });
    }
    getInstance() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.ledger) {
                this.ledger = new hw_app_cosmos_1.default((yield LedgerTransport.getTransport()));
            }
            return this.ledger;
        });
    }
    getAccountManager() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.accountManager) {
                this.accountManager = new AccountManager_js_1.default(yield this.getInstance());
            }
            return this.accountManager;
        });
    }
    refresh() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.ledger) {
                return new LedgerTransport();
            }
            this.ledger.transport.close();
            return new LedgerTransport();
        });
    }
}
exports.default = LedgerTransport;
