// @injectivelabs/wallet-ledger/dist/esm/strategy/LedgerCosmos/hw/AccountManager.js
import { DEFAULT_NUM_ADDRESSES_TO_FETCH } from '@injectivelabs/wallet-base';
export default class AccountManager {
    wallets = [];
    ledger;
    constructor(ledger) {
        this.ledger = ledger;
        this.wallets = [];
    }
    async getWallets(baseDerivationPath) {
        const { start, end } = this.getOffset();
        /**
         * 1. Wallets are not yet fetched at all,
         * 2. Wallets are not yet fetched for that offset
         */
        if (!this.hasWallets() || !this.hasWalletsInOffset(start)) {
            await this.getWalletsBasedOnIndex({
                start,
                end,
                baseDerivationPath,
            });
        }
        return this.wallets.slice(start, end);
    }
    getLedgerDerivationPathBasedOnType = ({ fullBaseDerivationPath, index, }) => {
        return `${fullBaseDerivationPath}/${index}'/0/0`;
    };
    async getWalletsBasedOnIndex({ start, end, baseDerivationPath, }) {
        for (let index = start; index < end; index += 1) {
            const path = this.getLedgerDerivationPathBasedOnType({
                fullBaseDerivationPath: baseDerivationPath,
                index,
            });
            const { address, publicKey } = await this.ledger.getAddress(path, 'inj');
            this.wallets.push({
                publicKey,
                baseDerivationPath,
                address: address.toLowerCase(),
                derivationPath: path,
            });
        }
    }
    hasWallets() {
        return this.wallets.length > 0;
    }
    hasWalletsInOffset(offset) {
        return this.wallets.length > offset;
    }
    getOffset() {
        const totalWallets = this.wallets.length;
        const nextBatchStart = totalWallets;
        const nextBatchEnd = totalWallets + DEFAULT_NUM_ADDRESSES_TO_FETCH;
        return {
            start: nextBatchStart,
            end: nextBatchEnd,
        };
    }
    hasWalletForAddress(address) {
        return (this.wallets.find((wallet) => wallet.address.toLowerCase() === address.toLowerCase()) !== undefined);
    }
    async getWalletForAddress(address) {
        return this.wallets.find((wallet) => wallet.address.toLowerCase() === address.toLowerCase());
    }
    reset() {
        this.wallets = [];
    }
}
