// @injectivelabs/wallet-ts/dist/cjs/types/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletAction = exports.WalletEventListener = exports.WalletDeviceType = exports.MagicProvider = exports.Wallet = void 0;
const exceptions_1 = require("@injectivelabs/exceptions");
var Wallet;
(function (Wallet) {
    Wallet["Leap"] = "leap";
    Wallet["Keplr"] = "keplr";
    Wallet["Ninji"] = "ninji";
    Wallet["Magic"] = "magic";
    Wallet["Torus"] = "torus";
    Wallet["Ledger"] = "ledger";
    Wallet["BitGet"] = "BitGet";
    Wallet["Trezor"] = "trezor";
    Wallet["Phantom"] = "phantom";
    Wallet["Metamask"] = "metamask";
    Wallet["OkxWallet"] = "okx-wallet";
    Wallet["TrustWallet"] = "trust-wallet";
    Wallet["PrivateKey"] = "private-key";
    Wallet["Cosmostation"] = "cosmostation";
    Wallet["LedgerCosmos"] = "ledger-cosmos";
    Wallet["LedgerLegacy"] = "ledger-legacy";
    Wallet["WalletConnect"] = "wallet-connect";
    Wallet["CosmostationEth"] = "cosmostation-eth";
})(Wallet || (exports.Wallet = Wallet = {}));
var MagicProvider;
(function (MagicProvider) {
    MagicProvider["Email"] = "email";
    MagicProvider["Apple"] = "apple";
    MagicProvider["Github"] = "github";
    MagicProvider["Google"] = "google";
    MagicProvider["Discord"] = "discord";
    MagicProvider["Twitter"] = "twitter";
    MagicProvider["Facebook"] = "facebook";
})(MagicProvider || (exports.MagicProvider = MagicProvider = {}));
var WalletDeviceType;
(function (WalletDeviceType) {
    WalletDeviceType["Mobile"] = "mobile";
    WalletDeviceType["Other"] = "other";
    WalletDeviceType["Browser"] = "browser";
    WalletDeviceType["Hardware"] = "hardware";
})(WalletDeviceType || (exports.WalletDeviceType = WalletDeviceType = {}));
var WalletEventListener;
(function (WalletEventListener) {
    WalletEventListener["AccountChange"] = "account-change";
    WalletEventListener["ChainIdChange"] = "chain-id-change";
})(WalletEventListener || (exports.WalletEventListener = WalletEventListener = {}));
exports.WalletAction = Object.assign({}, exceptions_1.WalletErrorActionModule);
