// @injectivelabs/wallet-ledger/dist/esm/strategy/LedgerCosmos/index.js
import { ErrorType, LedgerCosmosException, TransactionException, UnspecifiedErrorCode, WalletException, } from '@injectivelabs/exceptions';
import { toUtf8, TxGrpcApi, sortObjectByKeys, } from '@injectivelabs/sdk-ts';
import { WalletAction, WalletDeviceType, BaseConcreteStrategy, DEFAULT_BASE_DERIVATION_PATH, DEFAULT_ADDRESS_SEARCH_LIMIT, DEFAULT_NUM_ADDRESSES_TO_FETCH, } from '@injectivelabs/wallet-base';
import LedgerHW from './hw/index.js';
import { CosmosWalletException } from '@injectivelabs/exceptions';
export class LedgerCosmos extends BaseConcreteStrategy {
    baseDerivationPath;
    ledger;
    constructor(args) {
        super(args);
        this.baseDerivationPath = DEFAULT_BASE_DERIVATION_PATH;
        this.ledger = new LedgerHW();
    }
    async getWalletDeviceType() {
        return Promise.resolve(WalletDeviceType.Hardware);
    }
    async enable() {
        return Promise.resolve(true);
    }
    async disconnect() {
        this.ledger = await this.ledger.refresh();
    }
    async getAddresses() {
        const { baseDerivationPath } = this;
        try {
            const accountManager = await this.ledger.getAccountManager();
            const wallets = await accountManager.getWallets(baseDerivationPath);
            return wallets.map((k) => k.address);
        }
        catch (e) {
            throw new LedgerCosmosException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.GetAccounts,
            });
        }
    }
    async getSessionOrConfirm(address) {
        return Promise.resolve(`0x${Buffer.from(`Confirmation for ${address} at time: ${Date.now()}`).toString('hex')}`);
    }
    async sendEvmTransaction(_txData, _options) {
        throw new CosmosWalletException(new Error('sendEvmTransaction is not supported. LedgerCosmos only supports sending cosmos transactions'), {
            code: UnspecifiedErrorCode,
            context: WalletAction.SendEvmTransaction,
        });
    }
    async sendTransaction(transaction, options) {
        const { endpoints, txTimeout } = options;
        if (!endpoints) {
            throw new WalletException(new Error('You have to pass endpoints.grpc within the options for using LedgerCosmos wallet'));
        }
        const txApi = new TxGrpcApi(endpoints.grpc);
        const response = await txApi.broadcast(transaction, { txTimeout });
        if (response.code !== 0) {
            throw new TransactionException(new Error(response.rawLog), {
                code: UnspecifiedErrorCode,
                contextCode: response.code,
                contextModule: response.codespace,
            });
        }
        return response;
    }
    async signAminoCosmosTransaction(transaction) {
        try {
            const { derivationPath } = await this.getWalletForAddress(transaction.address);
            const ledger = await this.ledger.getInstance();
            const result = await ledger.sign(derivationPath, JSON.stringify(sortObjectByKeys(transaction.signDoc)));
            return {
                signed: undefined,
                signature: {
                    signature: Buffer.from(result.signature).toString('base64'),
                    pub_key: undefined,
                },
            };
        }
        catch (e) {
            throw new LedgerCosmosException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.SignTransaction,
            });
        }
    }
    async signCosmosTransaction(_transaction) {
        throw new CosmosWalletException(new Error('This wallet does not support signing using direct sign'), {
            code: UnspecifiedErrorCode,
            context: WalletAction.SendTransaction,
        });
    }
    async signEip712TypedData(_eip712Json, _address) {
        throw new CosmosWalletException(new Error('This wallet does not support signing Ethereum transactions'), {
            code: UnspecifiedErrorCode,
            context: WalletAction.SendTransaction,
        });
    }
    async signArbitrary(signer, data) {
        try {
            const { derivationPath } = await this.getWalletForAddress(signer);
            const ledger = await this.ledger.getInstance();
            const result = await ledger.sign(derivationPath, toUtf8(data));
            return Buffer.from(result.signature).toString('base64');
        }
        catch (e) {
            throw new LedgerCosmosException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.SignTransaction,
            });
        }
    }
    async getEthereumChainId() {
        throw new CosmosWalletException(new Error('getEthereumChainId is not supported on Keplr'), {
            code: UnspecifiedErrorCode,
            context: WalletAction.GetChainId,
        });
    }
    async getEvmTransactionReceipt(_txHash) {
        throw new CosmosWalletException(new Error('getEvmTransactionReceipt is not supported on Keplr'), {
            code: UnspecifiedErrorCode,
            context: WalletAction.GetEvmTransactionReceipt,
        });
    }
    async getPubKey(address) {
        if (!address) {
            throw new WalletException(new Error('You can only fetch PubKey corresponding to an address'));
        }
        const ledgerWalletInfo = await this.getWalletForAddress(address);
        return Buffer.from(ledgerWalletInfo.publicKey || '', 'hex').toString('base64');
    }
    async getWalletForAddress(address) {
        try {
            const { baseDerivationPath } = this;
            const accountManager = await this.ledger.getAccountManager();
            if (!accountManager.hasWalletForAddress(address)) {
                for (let i = 0; i < DEFAULT_ADDRESS_SEARCH_LIMIT / DEFAULT_NUM_ADDRESSES_TO_FETCH; i += 1) {
                    await accountManager.getWallets(baseDerivationPath);
                    if (accountManager.hasWalletForAddress(address)) {
                        return (await accountManager.getWalletForAddress(address));
                    }
                }
            }
            return (await accountManager.getWalletForAddress(address));
        }
        catch (e) {
            throw new LedgerCosmosException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.GetAccounts,
            });
        }
    }
}
