// @injectivelabs/wallet-evm/dist/cjs/strategy/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTrustWalletProvider = exports.getMetamaskProvider = exports.getRainbowProvider = exports.getPhantomProvider = exports.getBitGetProvider = exports.getOkxWalletProvider = void 0;
var Okx_js_1 = require("./Okx.js");
Object.defineProperty(exports, "getOkxWalletProvider", { enumerable: true, get: function () { return Okx_js_1.getOkxWalletProvider; } });
var bitget_js_1 = require("./bitget.js");
Object.defineProperty(exports, "getBitGetProvider", { enumerable: true, get: function () { return bitget_js_1.getBitGetProvider; } });
var phantom_js_1 = require("./phantom.js");
Object.defineProperty(exports, "getPhantomProvider", { enumerable: true, get: function () { return phantom_js_1.getPhantomProvider; } });
var rainbow_js_1 = require("./rainbow.js");
Object.defineProperty(exports, "getRainbowProvider", { enumerable: true, get: function () { return rainbow_js_1.getRainbowProvider; } });
var metamask_js_1 = require("./metamask.js");
Object.defineProperty(exports, "getMetamaskProvider", { enumerable: true, get: function () { return metamask_js_1.getMetamaskProvider; } });
var trustWallet_js_1 = require("./trustWallet.js");
Object.defineProperty(exports, "getTrustWalletProvider", { enumerable: true, get: function () { return trustWallet_js_1.getTrustWalletProvider; } });
