// @injectivelabs/wallet-base/dist/cjs/utils/wallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCosmosAminoOnlyWallet = exports.isEip712V2OnlyWallet = exports.isCosmosBrowserWallet = exports.isEvmBrowserWallet = exports.isCosmosWallet = exports.isEvmWallet = void 0;
const enums_js_1 = require("./../types/enums.js");
const isEvmWallet = (wallet) => [
    enums_js_1.Wallet.Magic,
    enums_js_1.Wallet.BitGet,
    enums_js_1.Wallet.Ledger,
    enums_js_1.Wallet.Phantom,
    enums_js_1.Wallet.Rainbow,
    enums_js_1.Wallet.Turnkey,
    enums_js_1.Wallet.Metamask,
    enums_js_1.Wallet.OkxWallet,
    enums_js_1.Wallet.PrivateKey,
    enums_js_1.Wallet.TrezorBip32,
    enums_js_1.Wallet.TrezorBip44,
    enums_js_1.Wallet.TrustWallet,
    enums_js_1.Wallet.LedgerLegacy,
    enums_js_1.Wallet.WalletConnect,
    enums_js_1.Wallet.CosmostationEth,
].includes(wallet);
exports.isEvmWallet = isEvmWallet;
const isCosmosWallet = (wallet) => !(0, exports.isEvmWallet)(wallet);
exports.isCosmosWallet = isCosmosWallet;
const isEvmBrowserWallet = (wallet) => [
    enums_js_1.Wallet.BitGet,
    enums_js_1.Wallet.Rainbow,
    enums_js_1.Wallet.Phantom,
    enums_js_1.Wallet.Turnkey,
    enums_js_1.Wallet.Metamask,
    enums_js_1.Wallet.OkxWallet,
    enums_js_1.Wallet.TrustWallet,
].includes(wallet);
exports.isEvmBrowserWallet = isEvmBrowserWallet;
const isCosmosBrowserWallet = (wallet) => [
    enums_js_1.Wallet.Leap,
    enums_js_1.Wallet.Ninji,
    enums_js_1.Wallet.Keplr,
    enums_js_1.Wallet.OWallet,
    enums_js_1.Wallet.Cosmostation,
].includes(wallet);
exports.isCosmosBrowserWallet = isCosmosBrowserWallet;
const isEip712V2OnlyWallet = (wallet) => [
    enums_js_1.Wallet.Magic,
    enums_js_1.Wallet.Metamask,
    enums_js_1.Wallet.Phantom,
    enums_js_1.Wallet.WalletConnect,
].includes(wallet);
exports.isEip712V2OnlyWallet = isEip712V2OnlyWallet;
const isCosmosAminoOnlyWallet = (wallet) => [enums_js_1.Wallet.LedgerCosmos].includes(wallet);
exports.isCosmosAminoOnlyWallet = isCosmosAminoOnlyWallet;
