// @keplr-wallet/proto-types/proto-types-gen/src/babylon/incentive/params.ts
/* eslint-disable */
import Long from "long";
import _m0 from "protobufjs/minimal";

export const protobufPackage = "babylon.incentive";

/**
 * Params defines the parameters for the module, including portions of rewards
 * distributed to each type of stakeholder. Note that sum of the portions should
 * be strictly less than 1 so that the rest will go to Comet
 * validators/delegations adapted from
 * https://github.com/cosmos/cosmos-sdk/blob/release/v0.47.x/proto/cosmos/distribution/v1beta1/distribution.proto
 */
export interface Params {
  /**
   * btc_staking_portion is the portion of rewards that goes to Finality
   * Providers/delegations NOTE: the portion of each Finality
   * Provider/delegation is calculated by using its voting power and finality
   * provider's commission
   */
  btcStakingPortion: string;
}

function createBaseParams(): Params {
  return { btcStakingPortion: "" };
}

export const Params = {
  encode(
    message: Params,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    if (message.btcStakingPortion !== "") {
      writer.uint32(10).string(message.btcStakingPortion);
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): Params {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseParams();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.btcStakingPortion = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): Params {
    return {
      btcStakingPortion: isSet(object.btcStakingPortion)
        ? String(object.btcStakingPortion)
        : "",
    };
  },

  toJSON(message: Params): unknown {
    const obj: any = {};
    message.btcStakingPortion !== undefined &&
      (obj.btcStakingPortion = message.btcStakingPortion);
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<Params>, I>>(object: I): Params {
    const message = createBaseParams();
    message.btcStakingPortion = object.btcStakingPortion ?? "";
    return message;
  },
};

type Builtin =
  | Date
  | Function
  | Uint8Array
  | string
  | number
  | boolean
  | undefined;

export type DeepPartial<T> = T extends Builtin
  ? T
  : T extends Array<infer U>
  ? Array<DeepPartial<U>>
  : T extends ReadonlyArray<infer U>
  ? ReadonlyArray<DeepPartial<U>>
  : T extends {}
  ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin
  ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & Record<
        Exclude<keyof I, KeysOfUnion<P>>,
        never
      >;

if (_m0.util.Long !== Long) {
  _m0.util.Long = Long as any;
  _m0.configure();
}

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
