// @injectivelabs/wallet-ts/dist/esm/strategies/wallet-strategy/strategies/WalletConnect.js
/* eslint-disable class-methods-use-this */
import { EthereumChainId } from '@injectivelabs/ts-types';
import { ErrorType, WalletException, MetamaskException, UnspecifiedErrorCode, TransactionException, } from '@injectivelabs/exceptions';
import { toUtf8, TxGrpcApi } from '@injectivelabs/sdk-ts';
import BaseConcreteStrategy from './Base.js';
import { WalletAction, WalletDeviceType, WalletEventListener, } from '../../../types/enums.js';
import { EthereumProvider, } from '@walletconnect/ethereum-provider';
const WalletConnectIds = {
    FireBlocks: '5864e2ced7c293ed18ac35e0db085c09ed567d67346ccb6f58a0327a75137489',
};
export default class WalletConnect extends BaseConcreteStrategy {
    provider;
    metadata;
    constructor(args) {
        super(args);
        this.metadata = args.metadata;
    }
    async getWalletDeviceType() {
        return Promise.resolve(WalletDeviceType.Browser);
    }
    async enable(args) {
        await this.connectWalletConnect(args?.topic);
        return Promise.resolve(true);
    }
    async disconnect() {
        if (this.listeners[WalletEventListener.AccountChange]) {
            const wc = await this.getConnectedWalletConnect();
            wc.removeListener('accountsChanged', this.listeners[WalletEventListener.AccountChange]);
        }
        if (this.listeners[WalletEventListener.ChainIdChange]) {
            const wc = await this.getConnectedWalletConnect();
            wc.removeListener('chainChanged', this.listeners[WalletEventListener.ChainIdChange]);
        }
        this.listeners = {};
        if (this.provider) {
            await this.provider.disconnect();
            this.provider = undefined;
        }
    }
    async getAddresses() {
        const wc = await this.getConnectedWalletConnect();
        try {
            return await wc.request({
                method: 'eth_requestAccounts',
            });
        }
        catch (e) {
            throw new MetamaskException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.GetAccounts,
            });
        }
    }
    async getSessionOrConfirm(_address) {
        const wc = await this.getConnectedWalletConnect();
        return wc.session?.topic || '';
    }
    async sendEthereumTransaction(transaction, _options) {
        const wc = await this.getConnectedWalletConnect();
        try {
            return await wc.request({
                method: 'eth_sendTransaction',
                params: [transaction],
            });
        }
        catch (e) {
            throw new MetamaskException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.SendEthereumTransaction,
            });
        }
    }
    async sendTransaction(transaction, options) {
        const { endpoints, txTimeout } = options;
        if (!endpoints) {
            throw new WalletException(new Error('You have to pass endpoints within the options for using Ethereum native wallets'));
        }
        const txApi = new TxGrpcApi(endpoints.grpc);
        const response = await txApi.broadcast(transaction, { txTimeout });
        if (response.code !== 0) {
            throw new TransactionException(new Error(response.rawLog), {
                code: UnspecifiedErrorCode,
                contextCode: response.code,
                contextModule: response.codespace,
            });
        }
        return response;
    }
    /** @deprecated */
    async signTransaction(eip712json, address) {
        return this.signEip712TypedData(eip712json, address);
    }
    async signEip712TypedData(eip712json, address) {
        const wc = await this.getConnectedWalletConnect();
        try {
            return await wc.request({
                method: 'eth_signTypedData_v4',
                params: [address, eip712json],
            });
        }
        catch (e) {
            throw new MetamaskException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.SignTransaction,
            });
        }
    }
    async signAminoCosmosTransaction(_transaction) {
        throw new WalletException(new Error('This wallet does not support signing Cosmos transactions'), {
            code: UnspecifiedErrorCode,
            type: ErrorType.WalletError,
            contextModule: WalletAction.SignTransaction,
        });
    }
    // eslint-disable-next-line class-methods-use-this
    async signCosmosTransaction(_transaction) {
        throw new WalletException(new Error('This wallet does not support signing Cosmos transactions'), {
            code: UnspecifiedErrorCode,
            type: ErrorType.WalletError,
            contextModule: WalletAction.SignTransaction,
        });
    }
    async signArbitrary(signer, data) {
        const wc = await this.getConnectedWalletConnect();
        try {
            const signature = await wc.request({
                method: 'personal_sign',
                params: [toUtf8(data), signer],
            });
            return signature;
        }
        catch (e) {
            throw new MetamaskException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.SignArbitrary,
            });
        }
    }
    async getEthereumChainId() {
        const wc = await this.getConnectedWalletConnect();
        try {
            return wc.request({ method: 'eth_chainId' });
        }
        catch (e) {
            throw new MetamaskException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.GetChainId,
            });
        }
    }
    async getEthereumTransactionReceipt(_txHash) {
        throw new WalletException(new Error('This wallet does not support awaiting Ethereum transactions'), {
            code: UnspecifiedErrorCode,
            type: ErrorType.WalletError,
            contextModule: WalletAction.GetEthereumTransactionReceipt,
        });
    }
    // eslint-disable-next-line class-methods-use-this
    async getPubKey() {
        throw new WalletException(new Error('You can only fetch PubKey from Cosmos native wallets'));
    }
    async onChainIdChanged(callback) {
        const wc = await this.getConnectedWalletConnect();
        this.listeners = {
            [WalletEventListener.ChainIdChange]: callback,
        };
        wc.on('chainChanged', callback);
    }
    async onAccountChange(callback) {
        const wc = await this.getConnectedWalletConnect();
        this.listeners = {
            [WalletEventListener.AccountChange]: callback,
        };
        wc.on('accountsChanged', (accounts) => callback(accounts[0]));
    }
    async getWalletConnect() {
        if (this.provider) {
            return this.provider;
        }
        if (!this.metadata) {
            throw new WalletException(new Error('Please provide metadata for WalletConnect'), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.GetAccounts,
            });
        }
        if (!this.metadata.projectId) {
            throw new WalletException(new Error('Please provide projectId alongside the metadata for WalletConnect'), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.GetAccounts,
            });
        }
        try {
            this.provider = await EthereumProvider.init({
                projectId: this.metadata.projectId,
                metadata: this.metadata
                    .metadata,
                showQrModal: true,
                optionalChains: this.ethereumChainId
                    ? [this.ethereumChainId]
                    : [EthereumChainId.Mainnet, EthereumChainId.Sepolia],
                qrModalOptions: {
                    explorerRecommendedWalletIds: [WalletConnectIds.FireBlocks],
                    explorerExcludedWalletIds: 'ALL',
                    mobileWallets: [],
                    walletImages: {
                        [WalletConnectIds.FireBlocks]: '/wallet-connect/fireblocks.webp',
                    },
                    desktopWallets: [
                        {
                            id: WalletConnectIds.FireBlocks,
                            name: 'Fireblocks',
                            links: {
                                native: 'fireblocks-wc://',
                                universal: 'https://console.fireblocks.io/v2/',
                            },
                        },
                    ],
                },
            });
            return this.provider;
        }
        catch (e) {
            throw new WalletException(new Error('WalletConnect not supported for this wallet'), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletError,
                contextModule: WalletAction.GetAccounts,
            });
        }
    }
    async getConnectedWalletConnect() {
        if (!this.provider) {
            await this.getWalletConnect();
        }
        if (!this.provider?.connected) {
            await this.enable();
        }
        return this.provider;
    }
    async connectWalletConnect(topic) {
        if (this.provider && this.provider.connected) {
            return;
        }
        const wc = await this.getWalletConnect();
        await wc.connect({
            ...(topic && { pairingTopic: topic }),
        });
    }
}
