// @injectivelabs/wallet-base/dist/esm/types/enums.js
import { WalletErrorActionModule } from '@injectivelabs/exceptions';
export var BroadcastMode;
(function (BroadcastMode) {
    /** Return after tx commit */
    BroadcastMode["Block"] = "block";
    /** Return after CheckTx */
    BroadcastMode["Sync"] = "sync";
    /** Return right away */
    BroadcastMode["Async"] = "async";
})(BroadcastMode || (BroadcastMode = {}));
export var Wallet;
(function (Wallet) {
    Wallet["Leap"] = "leap";
    Wallet["Keplr"] = "keplr";
    Wallet["Ninji"] = "ninji";
    Wallet["Magic"] = "magic";
    Wallet["Ledger"] = "ledger";
    Wallet["BitGet"] = "BitGet";
    Wallet["OWallet"] = "owallet";
    Wallet["Phantom"] = "phantom";
    Wallet["Rainbow"] = "rainbow";
    Wallet["Turnkey"] = "turnkey";
    Wallet["Metamask"] = "metamask";
    Wallet["OkxWallet"] = "okx-wallet";
    Wallet["PrivateKey"] = "private-key";
    Wallet["TrustWallet"] = "trust-wallet";
    Wallet["TrezorBip32"] = "trezor-bip32";
    Wallet["TrezorBip44"] = "trezor-bip44";
    Wallet["Cosmostation"] = "cosmostation";
    Wallet["LedgerCosmos"] = "ledger-cosmos";
    Wallet["LedgerLegacy"] = "ledger-legacy";
    Wallet["WalletConnect"] = "wallet-connect";
    Wallet["CosmostationEth"] = "cosmostation-eth";
})(Wallet || (Wallet = {}));
export var MagicProvider;
(function (MagicProvider) {
    MagicProvider["Email"] = "email";
    MagicProvider["Apple"] = "apple";
    MagicProvider["Github"] = "github";
    MagicProvider["Google"] = "google";
    MagicProvider["Discord"] = "discord";
    MagicProvider["Twitter"] = "twitter";
    MagicProvider["Facebook"] = "facebook";
})(MagicProvider || (MagicProvider = {}));
export var WalletDeviceType;
(function (WalletDeviceType) {
    WalletDeviceType["Mobile"] = "mobile";
    WalletDeviceType["Other"] = "other";
    WalletDeviceType["Browser"] = "browser";
    WalletDeviceType["Hardware"] = "hardware";
})(WalletDeviceType || (WalletDeviceType = {}));
export var WalletEventListener;
(function (WalletEventListener) {
    WalletEventListener["AccountChange"] = "account-change";
    WalletEventListener["ChainIdChange"] = "chain-id-change";
})(WalletEventListener || (WalletEventListener = {}));
export const WalletAction = { ...WalletErrorActionModule };
