// @injectivelabs/wallet-ts/dist/esm/strategies/cosmos-wallet-strategy/CosmosWalletStrategy.js
import { GeneralException } from '@injectivelabs/exceptions';
import { Wallet } from '../../types/enums.js';
import Keplr from './strategies/Keplr.js';
import Leap from './strategies/Leap.js';
import Cosmostation from './strategies/Cosmostation.js';
import Ninji from './strategies/Ninji.js';
import { isCosmosWallet } from '../wallet-strategy/utils.js';
export const cosmosWallets = [
    Wallet.Keplr,
    Wallet.Leap,
    Wallet.Cosmostation,
    Wallet.Ninji,
];
const createWallet = ({ wallet, args, }) => {
    switch (wallet) {
        case Wallet.Keplr:
            return new Keplr({ ...args });
        case Wallet.Leap:
            return new Leap({ ...args });
        case Wallet.Cosmostation:
            return new Cosmostation({ ...args });
        case Wallet.Ninji:
            return new Ninji({ ...args });
        default:
            throw new GeneralException(new Error(`The ${wallet} concrete wallet strategy is not supported`));
    }
};
const createWallets = (args) => cosmosWallets.reduce((strategies, wallet) => {
    if (strategies[wallet]) {
        return strategies;
    }
    strategies[wallet] = createWallet({ wallet, args });
    return strategies;
}, {});
export default class CosmosWalletStrategy {
    strategies;
    wallet;
    constructor(args) {
        this.strategies = createWallets(args);
        this.wallet = args.wallet || Wallet.Keplr;
    }
    getWallet() {
        return this.wallet;
    }
    setWallet(wallet) {
        this.wallet = isCosmosWallet(wallet) ? wallet : Wallet.Keplr;
    }
    getStrategy() {
        if (!this.strategies[this.wallet]) {
            throw new GeneralException(new Error(`Wallet ${this.wallet} is not enabled/available!`));
        }
        return this.strategies[this.wallet];
    }
    getWalletDeviceType() {
        return this.getStrategy().getWalletDeviceType();
    }
    getPubKey() {
        return this.getStrategy().getPubKey();
    }
    enable() {
        return this.getStrategy().enable();
    }
    getAddresses() {
        return this.getStrategy().getAddresses();
    }
    async enableAndGetAddresses() {
        await this.getStrategy().enable();
        return await this.getStrategy().getAddresses();
    }
    async sendTransaction(tx, options) {
        return this.getStrategy().sendTransaction(tx, options);
    }
    async signTransaction(transaction) {
        return this.getStrategy().signTransaction(transaction);
    }
    async signAminoTransaction(transaction) {
        return this.getStrategy().signAminoTransaction(transaction);
    }
}
