// @ledgerhq/cryptoassets-evm-signatures/lib-es/crypto-assets-importer/fetch/index.js
import { getEnv } from "@ledgerhq/live-env";
import axios, { AxiosError } from "axios";
const LEDGER_COUNTERVALUES_API = getEnv("LEDGER_COUNTERVALUES_API");
export const fetchTokensFromCDN = async (filename) => {
    try {
        const { data, headers } = await axios.get(`${getEnv("DYNAMIC_CAL_BASE_URL")}/${filename}`);
        return [data, headers.etag];
    }
    catch (err) {
        const error = err;
        throw new Error(`${error.message} ${getEnv("DYNAMIC_CAL_BASE_URL")}/${filename}`);
    }
};
export const fetchTokensFromCALService = async (chainDetails, output, etag, next) => {
    try {
        const { data, headers } = await axios.get(`${getEnv("CAL_SERVICE_URL")}/v1/tokens`, {
            params: {
                ...(next?.cursor ? { cursor: next.cursor } : {}),
                chain_id: chainDetails.chainId,
                standard: chainDetails.standard,
                blockchain_name: chainDetails.blockchain_name,
                output: output.join(),
            },
            headers: etag
                ? {
                    "If-None-Match": etag,
                }
                : undefined,
        });
        const cursor = headers["x-ledger-next"];
        if (cursor) {
            return fetchTokensFromCALService({ chainId: chainDetails.chainId, standard: chainDetails.standard }, output, etag, {
                tokens: next?.tokens ? [...next.tokens, ...data] : data,
                cursor,
            });
        }
        const hash = headers["etag"];
        return next?.tokens ? { tokens: [...next.tokens, ...data], hash } : { tokens: data, hash };
    }
    catch (err) {
        if (err instanceof AxiosError) {
            if (err.response?.status === 304) {
                throw err;
            }
            throw new Error(`${err?.message}: ${err?.config?.url}`);
        }
        throw err;
    }
};
export const fetchTokensOrderedByMarketCap = async () => {
    const { data } = await axios.get(`${LEDGER_COUNTERVALUES_API}/v3/supported/crypto`);
    return { tokens: data };
};
