// @injectivelabs/wallet-ts/dist/cjs/strategies/cosmos-wallet-strategy/CosmosWalletStrategy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cosmosWallets = void 0;
const exceptions_1 = require("@injectivelabs/exceptions");
const enums_js_1 = require("../../types/enums.js");
const Keplr_js_1 = __importDefault(require("./strategies/Keplr.js"));
const Leap_js_1 = __importDefault(require("./strategies/Leap.js"));
const Cosmostation_js_1 = __importDefault(require("./strategies/Cosmostation.js"));
const Ninji_js_1 = __importDefault(require("./strategies/Ninji.js"));
const utils_js_1 = require("../wallet-strategy/utils.js");
exports.cosmosWallets = [
    enums_js_1.Wallet.Keplr,
    enums_js_1.Wallet.Leap,
    enums_js_1.Wallet.Cosmostation,
    enums_js_1.Wallet.Ninji,
];
const createWallet = ({ wallet, args, }) => {
    switch (wallet) {
        case enums_js_1.Wallet.Keplr:
            return new Keplr_js_1.default(Object.assign({}, args));
        case enums_js_1.Wallet.Leap:
            return new Leap_js_1.default(Object.assign({}, args));
        case enums_js_1.Wallet.Cosmostation:
            return new Cosmostation_js_1.default(Object.assign({}, args));
        case enums_js_1.Wallet.Ninji:
            return new Ninji_js_1.default(Object.assign({}, args));
        default:
            throw new exceptions_1.GeneralException(new Error(`The ${wallet} concrete wallet strategy is not supported`));
    }
};
const createWallets = (args) => exports.cosmosWallets.reduce((strategies, wallet) => {
    if (strategies[wallet]) {
        return strategies;
    }
    strategies[wallet] = createWallet({ wallet, args });
    return strategies;
}, {});
class CosmosWalletStrategy {
    constructor(args) {
        this.strategies = createWallets(args);
        this.wallet = args.wallet || enums_js_1.Wallet.Keplr;
    }
    getWallet() {
        return this.wallet;
    }
    setWallet(wallet) {
        this.wallet = (0, utils_js_1.isCosmosWallet)(wallet) ? wallet : enums_js_1.Wallet.Keplr;
    }
    getStrategy() {
        if (!this.strategies[this.wallet]) {
            throw new exceptions_1.GeneralException(new Error(`Wallet ${this.wallet} is not enabled/available!`));
        }
        return this.strategies[this.wallet];
    }
    getWalletDeviceType() {
        return this.getStrategy().getWalletDeviceType();
    }
    getPubKey() {
        return this.getStrategy().getPubKey();
    }
    enable() {
        return this.getStrategy().enable();
    }
    getAddresses() {
        return this.getStrategy().getAddresses();
    }
    enableAndGetAddresses() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.getStrategy().enable();
            return yield this.getStrategy().getAddresses();
        });
    }
    sendTransaction(tx, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.getStrategy().sendTransaction(tx, options);
        });
    }
    signTransaction(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.getStrategy().signTransaction(transaction);
        });
    }
    signAminoTransaction(transaction) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.getStrategy().signAminoTransaction(transaction);
        });
    }
}
exports.default = CosmosWalletStrategy;
