// @injectivelabs/wallet-ts/dist/esm/strategies/cosmos-wallet-strategy/strategies/Leap.js
/* eslint-disable class-methods-use-this */
import { CosmosChainId } from '@injectivelabs/ts-types';
import { ErrorType, UnspecifiedErrorCode, CosmosWalletException, TransactionException, } from '@injectivelabs/exceptions';
import { waitTxBroadcasted, createTxRawFromSigResponse, createSignDocFromTransaction, } from '@injectivelabs/sdk-ts';
import { LeapWallet } from '../../../utils/wallets/leap/index.js';
import { WalletAction, WalletDeviceType } from '../../../types/enums.js';
import { createCosmosSignDocFromSignDoc } from '../../../utils/cosmos.js';
export default class Leap {
    chainId;
    leapWallet;
    constructor(args) {
        this.chainId = args.chainId || CosmosChainId.Injective;
        this.leapWallet = new LeapWallet(args.chainId);
    }
    async getWalletDeviceType() {
        const leapWallet = this.getLeapWallet();
        const key = await leapWallet.getKey();
        return key.isNanoLedger
            ? Promise.resolve(WalletDeviceType.Hardware)
            : Promise.resolve(WalletDeviceType.Browser);
    }
    async enable() {
        return await LeapWallet.isChainIdSupported(this.chainId);
    }
    async getAddresses() {
        const leapWallet = this.getLeapWallet();
        try {
            const accounts = await leapWallet.getAccounts();
            return accounts.map((account) => account.address);
        }
        catch (e) {
            if (e instanceof CosmosWalletException) {
                throw e;
            }
            throw new CosmosWalletException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                context: WalletAction.GetAccounts,
            });
        }
    }
    async sendTransaction(transaction, options) {
        const { leapWallet } = this;
        const txRaw = createTxRawFromSigResponse(transaction);
        if (!options.endpoints) {
            throw new CosmosWalletException(new Error('You have to pass endpoints within the options to broadcast transaction'));
        }
        try {
            const txHash = await leapWallet.broadcastTx(txRaw);
            return await waitTxBroadcasted(txHash, options);
        }
        catch (e) {
            throw new TransactionException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                context: WalletAction.SendTransaction,
            });
        }
    }
    async signTransaction(transaction) {
        const leapWallet = this.getLeapWallet();
        const signer = await leapWallet.getOfflineSigner();
        const signDoc = createSignDocFromTransaction(transaction);
        try {
            return signer.signDirect(transaction.address, createCosmosSignDocFromSignDoc(signDoc));
        }
        catch (e) {
            if (e instanceof TransactionException) {
                throw e;
            }
            throw new CosmosWalletException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                context: WalletAction.SendTransaction,
            });
        }
    }
    async signAminoTransaction(_transaction) {
        throw new CosmosWalletException(new Error('signAminoTransaction not supported on Leap'));
    }
    async getPubKey() {
        const leapWallet = this.getLeapWallet();
        try {
            const key = await leapWallet.getKey();
            return Buffer.from(key.pubKey).toString('base64');
        }
        catch (e) {
            throw new CosmosWalletException(new Error(e.message), {
                code: UnspecifiedErrorCode,
                context: WalletAction.GetAccounts,
            });
        }
    }
    getLeapWallet() {
        const { leapWallet } = this;
        if (!leapWallet) {
            throw new CosmosWalletException(new Error('Please install the Leap wallet extension'), {
                code: UnspecifiedErrorCode,
                type: ErrorType.WalletNotInstalledError,
                context: WalletAction.SignTransaction,
            });
        }
        return leapWallet;
    }
}
