// @ledgerhq/evm-tools/lib/selectors/dappSelectors.js
"use strict";
// This selectors will have duplicate values, different signatures can have the same function name but it will have different parameters.
/* eslint-disable @typescript-eslint/no-duplicate-enum-values */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZKSYNC_CLEAR_SIGNED_SELECTORS = exports.POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS = exports.POLYGON_CLEAR_SIGNED_SELECTORS = exports.OPTIMISM_CLEAR_SIGNED_SELECTORS = exports.FANTOM_CLEAR_SIGNED_SELECTORS = exports.ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS = exports.ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS = exports.ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS = exports.ETHEREUM_CLEAR_SIGNED_SELECTORS = exports.BSC_CLEAR_SIGNED_SELECTORS = exports.BASE_CLEAR_SIGNED_SELECTORS = exports.AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS = exports.ARBITRUM_CLEAR_SIGNED_SELECTORS = void 0;
var ARBITRUM_CLEAR_SIGNED_SELECTORS;
(function (ARBITRUM_CLEAR_SIGNED_SELECTORS) {
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x016cba5f"] = "execute";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x0502b1c5"] = "unoswap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x0b86a4c1"] = "swapOnUniswapV2Fork";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x0d5f0e3b"] = "uniswapV3SwapTo";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x12a0ddc7"] = "PMMV2Swap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x12aa3caf"] = "swap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x2298207a"] = "simpleBuy";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x2521b930"] = "uniswapV3SwapToWithPermit";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x259f8d1a"] = "withdraw";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x2e95b6c8"] = "unoswap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x3c15fd91"] = "unoswapToWithPermit";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x3d21e25a"] = "swapBridgeToV2";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x3eca9c0a"] = "fillOrderRFQ";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x46c67b6d"] = "megaSwap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x4cc4a27b"] = "fillOrderRFQToWithPermit";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x4d447995"] = "deposit";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x54840d1a"] = "swapOnUniswap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x54e3f31b"] = "SimpleSwap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x617ba037"] = "supply";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x64466805"] = "swapOnZeroXv4";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x65ae14b5"] = "deleverageWithdraw";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x69328dec"] = "withdraw";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x6e553f65"] = "deposit";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x70ccbd31"] = "fillOrderRFQToWithPermit";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x7c025200"] = "swap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x81033120"] = "swapOnZeroXv2";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x84bd6d29"] = "clipperSwap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x935fb84b"] = "buyOnUniswap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x972250fe"] = "bridgeToV2";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0x9871efa4"] = "unxswapByOrderId";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xa1251d75"] = "unoswapWithPermit";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xa2922622"] = "withdrawWithSymbolCheck";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xa6886da9"] = "directUniV3Swap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xa94e78ef"] = "multiSwap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xaa2daba6"] = "depositWithSymbolCheck";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xb460af94"] = "withdraw";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xb80c2f09"] = "smartSwapByOrderId";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xbc80f1a8"] = "uniswapV3SwapTo";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xc03786b0"] = "buyOnUniswapFork";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xc805a666"] = "clipperSwapToWithPermit";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xd0a3b665"] = "fillOrderRFQ";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xe449022e"] = "uniswapV3Swap";
    ARBITRUM_CLEAR_SIGNED_SELECTORS["0xf5661034"] = "swapOnUniswapFork";
})(ARBITRUM_CLEAR_SIGNED_SELECTORS || (exports.ARBITRUM_CLEAR_SIGNED_SELECTORS = ARBITRUM_CLEAR_SIGNED_SELECTORS = {}));
var AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS;
(function (AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS) {
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x016cba5f"] = "execute";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x0d10d32c"] = "redeemOverdueShares";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x0d5f0e3b"] = "uniswapV3SwapTo";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x0f7e2048"] = "redeemOverdueShares";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x12a0ddc7"] = "PMMV2Swap";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x3d21e25a"] = "swapBridgeToV2";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x5bcb2fc6"] = "submit";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x617ba037"] = "supply";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x69328dec"] = "withdraw";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x972250fe"] = "bridgeToV2";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0x9871efa4"] = "unxswapByOrderId";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0xb80c2f09"] = "smartSwapByOrderId";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0xbe040fb0"] = "redeem";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0xc9d2ff9d"] = "requestUnlock";
    AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS["0xdb006a75"] = "redeem";
})(AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS || (exports.AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS = AVALANCHE_C_CHAIN_CLEAR_SIGNED_SELECTORS = {}));
var BASE_CLEAR_SIGNED_SELECTORS;
(function (BASE_CLEAR_SIGNED_SELECTORS) {
    BASE_CLEAR_SIGNED_SELECTORS["0x016cba5f"] = "execute";
    BASE_CLEAR_SIGNED_SELECTORS["0x0502b1c5"] = "unoswap";
    BASE_CLEAR_SIGNED_SELECTORS["0x095ea7b3"] = "approve";
    BASE_CLEAR_SIGNED_SELECTORS["0x0b86a4c1"] = "swapOnUniswapV2Fork";
    BASE_CLEAR_SIGNED_SELECTORS["0x0d5f0e3b"] = "uniswapV3SwapTo";
    BASE_CLEAR_SIGNED_SELECTORS["0x12a0ddc7"] = "PMMV2Swap";
    BASE_CLEAR_SIGNED_SELECTORS["0x12aa3caf"] = "swap";
    BASE_CLEAR_SIGNED_SELECTORS["0x20b76e81"] = "repay";
    BASE_CLEAR_SIGNED_SELECTORS["0x2298207a"] = "simpleBuy";
    BASE_CLEAR_SIGNED_SELECTORS["0x238d6579"] = "supplyCollateral";
    BASE_CLEAR_SIGNED_SELECTORS["0x3c15fd91"] = "unoswapToWithPermit";
    BASE_CLEAR_SIGNED_SELECTORS["0x3d21e25a"] = "swapBridgeToV2";
    BASE_CLEAR_SIGNED_SELECTORS["0x3eca9c0a"] = "fillOrderRFQ";
    BASE_CLEAR_SIGNED_SELECTORS["0x46c67b6d"] = "megaSwap";
    BASE_CLEAR_SIGNED_SELECTORS["0x50d8cd4b"] = "borrow";
    BASE_CLEAR_SIGNED_SELECTORS["0x54840d1a"] = "swapOnUniswap";
    BASE_CLEAR_SIGNED_SELECTORS["0x54e3f31b"] = "SimpleSwap";
    BASE_CLEAR_SIGNED_SELECTORS["0x5c2bea49"] = "withdraw";
    BASE_CLEAR_SIGNED_SELECTORS["0x64466805"] = "swapOnZeroXv4";
    BASE_CLEAR_SIGNED_SELECTORS["0x6e553f65"] = "deposit";
    BASE_CLEAR_SIGNED_SELECTORS["0x70ccbd31"] = "fillOrderRFQToWithPermit";
    BASE_CLEAR_SIGNED_SELECTORS["0x8069218f"] = "setAuthorizationWithSig";
    BASE_CLEAR_SIGNED_SELECTORS["0x81033120"] = "swapOnZeroXv2";
    BASE_CLEAR_SIGNED_SELECTORS["0x833947fd"] = "reallocateTo";
    BASE_CLEAR_SIGNED_SELECTORS["0x84bd6d29"] = "clipperSwap";
    BASE_CLEAR_SIGNED_SELECTORS["0x8720316d"] = "withdrawCollateral";
    BASE_CLEAR_SIGNED_SELECTORS["0x8c1358a2"] = "createMarket";
    BASE_CLEAR_SIGNED_SELECTORS["0x935fb84b"] = "buyOnUniswap";
    BASE_CLEAR_SIGNED_SELECTORS["0x94bf804d"] = "mint";
    BASE_CLEAR_SIGNED_SELECTORS["0x972250fe"] = "bridgeToV2";
    BASE_CLEAR_SIGNED_SELECTORS["0x9871efa4"] = "unxswapByOrderId";
    BASE_CLEAR_SIGNED_SELECTORS["0xa6886da9"] = "directUniV3Swap";
    BASE_CLEAR_SIGNED_SELECTORS["0xa94e78ef"] = "multiSwap";
    BASE_CLEAR_SIGNED_SELECTORS["0xa99aad89"] = "supply";
    BASE_CLEAR_SIGNED_SELECTORS["0xb460af94"] = "withdraw";
    BASE_CLEAR_SIGNED_SELECTORS["0xb80c2f09"] = "smartSwapByOrderId";
    BASE_CLEAR_SIGNED_SELECTORS["0xba087652"] = "redeem";
    BASE_CLEAR_SIGNED_SELECTORS["0xc03786b0"] = "buyOnUniswapFork";
    BASE_CLEAR_SIGNED_SELECTORS["0xc805a666"] = "clipperSwapToWithPermit";
    BASE_CLEAR_SIGNED_SELECTORS["0xe0232b42"] = "flashLoan";
    BASE_CLEAR_SIGNED_SELECTORS["0xeecea000"] = "setAuthorization";
    BASE_CLEAR_SIGNED_SELECTORS["0xf5661034"] = "swapOnUniswapFork";
})(BASE_CLEAR_SIGNED_SELECTORS || (exports.BASE_CLEAR_SIGNED_SELECTORS = BASE_CLEAR_SIGNED_SELECTORS = {}));
var BSC_CLEAR_SIGNED_SELECTORS;
(function (BSC_CLEAR_SIGNED_SELECTORS) {
    BSC_CLEAR_SIGNED_SELECTORS["0x016cba5f"] = "execute";
    BSC_CLEAR_SIGNED_SELECTORS["0x0863b7ac"] = "SwapOnUniswapFork";
    BSC_CLEAR_SIGNED_SELECTORS["0x0d57afa6"] = "requestWithdraw";
    BSC_CLEAR_SIGNED_SELECTORS["0x0d5f0e3b"] = "uniswapV3SwapTo";
    BSC_CLEAR_SIGNED_SELECTORS["0x12a0ddc7"] = "PMMV2Swap";
    BSC_CLEAR_SIGNED_SELECTORS["0x1e6d24c2"] = "dodoSwapV2TokenToETH";
    BSC_CLEAR_SIGNED_SELECTORS["0x2298207a"] = "simpleBuy";
    BSC_CLEAR_SIGNED_SELECTORS["0x23b872dd"] = "transferFrom";
    BSC_CLEAR_SIGNED_SELECTORS["0x26ef699d"] = "sendToken";
    BSC_CLEAR_SIGNED_SELECTORS["0x2e1a7d4d"] = "withdraw";
    BSC_CLEAR_SIGNED_SELECTORS["0x2e95b6c8"] = "unoswap";
    BSC_CLEAR_SIGNED_SELECTORS["0x33635226"] = "BuyOnUniswapFork";
    BSC_CLEAR_SIGNED_SELECTORS["0x3c659741"] = "bridgeCall";
    BSC_CLEAR_SIGNED_SELECTORS["0x3d21e25a"] = "swapBridgeToV2";
    BSC_CLEAR_SIGNED_SELECTORS["0x3f2e5fc3"] = "sendNative";
    BSC_CLEAR_SIGNED_SELECTORS["0x46c67b6d"] = "megaSwap";
    BSC_CLEAR_SIGNED_SELECTORS["0x4d99dd16"] = "undelegate";
    BSC_CLEAR_SIGNED_SELECTORS["0x5028bb95"] = "dodoSwapV2ETHToToken";
    BSC_CLEAR_SIGNED_SELECTORS["0x51227094"] = "processOutputOrders";
    BSC_CLEAR_SIGNED_SELECTORS["0x54840d1a"] = "swapOnUniswap";
    BSC_CLEAR_SIGNED_SELECTORS["0x54bacd13"] = "externalSwap";
    BSC_CLEAR_SIGNED_SELECTORS["0x54e3f31b"] = "SimpleSwap";
    BSC_CLEAR_SIGNED_SELECTORS["0x58b9d179"] = "SwapOnUniswap";
    BSC_CLEAR_SIGNED_SELECTORS["0x59491871"] = "redelegate";
    BSC_CLEAR_SIGNED_SELECTORS["0x64466805"] = "swapOnZeroXv4";
    BSC_CLEAR_SIGNED_SELECTORS["0x6d9634b7"] = "releaseTokens";
    BSC_CLEAR_SIGNED_SELECTORS["0x7617b389"] = "mixSwap";
    BSC_CLEAR_SIGNED_SELECTORS["0x7c025200"] = "swap";
    BSC_CLEAR_SIGNED_SELECTORS["0x81033120"] = "swapOnZeroXv2";
    BSC_CLEAR_SIGNED_SELECTORS["0x81791788"] = "dodoMutliSwap";
    BSC_CLEAR_SIGNED_SELECTORS["0x8ca3bf68"] = "callBridgeCall";
    BSC_CLEAR_SIGNED_SELECTORS["0x8f00eccb"] = "MultiSwap";
    BSC_CLEAR_SIGNED_SELECTORS["0x90e1aa69"] = "processInputOrders";
    BSC_CLEAR_SIGNED_SELECTORS["0x935fb84b"] = "buyOnUniswap";
    BSC_CLEAR_SIGNED_SELECTORS["0x972250fe"] = "bridgeToV2";
    BSC_CLEAR_SIGNED_SELECTORS["0x982ef0a7"] = "delegate";
    BSC_CLEAR_SIGNED_SELECTORS["0x9871efa4"] = "unxswapByOrderId";
    BSC_CLEAR_SIGNED_SELECTORS["0x9ddb511a"] = "delegate";
    BSC_CLEAR_SIGNED_SELECTORS["0xa21a9280"] = "withdraw";
    BSC_CLEAR_SIGNED_SELECTORS["0xa27e8b6b"] = "SimpleBuy";
    BSC_CLEAR_SIGNED_SELECTORS["0xa378534b"] = "create";
    BSC_CLEAR_SIGNED_SELECTORS["0xa5977fbb"] = "send";
    BSC_CLEAR_SIGNED_SELECTORS["0xa6886da9"] = "directUniV3Swap";
    BSC_CLEAR_SIGNED_SELECTORS["0xa94e78ef"] = "multiSwap";
    BSC_CLEAR_SIGNED_SELECTORS["0xaa7415f5"] = "transferOut";
    BSC_CLEAR_SIGNED_SELECTORS["0xaad3ec96"] = "claim";
    BSC_CLEAR_SIGNED_SELECTORS["0xb80c2f09"] = "smartSwapByOrderId";
    BSC_CLEAR_SIGNED_SELECTORS["0xbba9b10c"] = "destroy";
    BSC_CLEAR_SIGNED_SELECTORS["0xc03786b0"] = "buyOnUniswapFork";
    BSC_CLEAR_SIGNED_SELECTORS["0xcfc0afeb"] = "SimpleSwap";
    BSC_CLEAR_SIGNED_SELECTORS["0xd0e30db0"] = "deposit";
    BSC_CLEAR_SIGNED_SELECTORS["0xde790c7e"] = "burn";
    BSC_CLEAR_SIGNED_SELECTORS["0xec1d21dd"] = "MegaSwap";
    BSC_CLEAR_SIGNED_SELECTORS["0xf35af1f8"] = "callBridge";
    BSC_CLEAR_SIGNED_SELECTORS["0xf5661034"] = "swapOnUniswapFork";
    BSC_CLEAR_SIGNED_SELECTORS["0xf8444436"] = "claimWithdrawal";
    BSC_CLEAR_SIGNED_SELECTORS["0xf87dc1b7"] = "dodoSwapV2TokenToToken";
    BSC_CLEAR_SIGNED_SELECTORS["0xf9355f72"] = "BuyOnUniswap";
    BSC_CLEAR_SIGNED_SELECTORS["0xf95a49eb"] = "Buy";
})(BSC_CLEAR_SIGNED_SELECTORS || (exports.BSC_CLEAR_SIGNED_SELECTORS = BSC_CLEAR_SIGNED_SELECTORS = {}));
var ETHEREUM_CLEAR_SIGNED_SELECTORS;
(function (ETHEREUM_CLEAR_SIGNED_SELECTORS) {
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x00f714ce"] = "withdraw_to";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x016cba5f"] = "execute";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x01d523b6"] = "updateStateAndDepositAndMintOsToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x026e402b"] = "delegate";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x0502b1c5"] = "unoswap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x066055e0"] = "burnOsToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x074ee446"] = "mint";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x084fd9b4"] = "claimFromVault";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x0863b7ac"] = "swapOnUniswapFork";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x08dc9f42"] = "mint";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x095ea7b3"] = "approve";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x0968f264"] = "withdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x0b86a4c1"] = "swapOnUniswapV2Fork";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x0c0a769b"] = "supply";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x0d5f0e3b"] = "uniswapV3SwapTo";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x0dd8dd02"] = "queueWithdrawals";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x0e752702"] = "repayBorrow";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x0ffab6c2"] = "batchWithdrawCLFee";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x10f13a8c"] = "SetText";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x1249c58b"] = "mint";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x12a0ddc7"] = "PMMV2Swap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x12aa3caf"] = "swap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x148e23ce"] = "claimRewards";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x15373e3d"] = "castVote";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x1564da79"] = "submitPOL";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x16dc3ace"] = "swapExactAmountOut";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x1896f70a"] = "setResolver";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x18f72950"] = "updateStateAndDeposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x19aa6257"] = "requestWithdrawalsWstETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x1a7ff553"] = "updateState";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x1c4b774b"] = "getReward";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x1da649cf"] = "repay";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x1e6d24c2"] = "dodoSwapV2TokenToETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x1f7ec122"] = "requestWithdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x201b9eb5"] = "mintOsToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x224199c2"] = "ProveAndClaimWithResolver";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2298207a"] = "simpleBuy";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x22a775b6"] = "mintAndTransfer";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x23463624"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x23b872dd"] = "transferFrom";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x23e103a8"] = "withdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x242eba0e"] = "deletePendingValidators";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x24d83b79"] = "addToPerpetual";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2521b930"] = "uniswapV3SwapToWithPermit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x258d3c33"] = "redeemYield";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x259f8d1a"] = "withdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2608f818"] = "repayBorrowBehalf";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x26ef699d"] = "sendToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x27050d1f"] = "createToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2775d01c"] = "withdrawInstantly";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2aba2aeb"] = "addToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2ba03a79"] = "withdrawCLFee";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2c4b04fa"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2d8918d7"] = "addLiquidity";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2def6620"] = "unstake";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2e17de78"] = "unstake";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2e1a7d4d"] = "withdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2e2d2984"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x2e95b6c8"] = "unoswap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x304e6ade"] = "SetContentHash";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3111e7b3"] = "claimRewards";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x31807e42"] = "depositYieldToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x32ca494b"] = "withdrawWstETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x32f73258"] = "updateOperatorControllingAddress";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x33404396"] = "completeQueuedWithdrawals";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x33635226"] = "buyOnUniswapFork";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x379607f5"] = "claim";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x388bbfd3"] = "redeem";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3a4b66f1"] = "stake";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3c15fd91"] = "unoswapToWithPermit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3c659741"] = "bridgeCall";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3ccfd60b"] = "withdraw_all";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3d13f874"] = "claim";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3d18b912"] = "getReward";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3d21e25a"] = "swapBridgeToV2";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3db397c6"] = "__ERC721Rarible_init";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3e200d4b"] = "stakeRPL";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3eca9c0a"] = "fillOrderRFQ";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x3f2e5fc3"] = "sendNative";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x42966c68"] = "burn";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x439370b1"] = "depositEth";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x43e82a79"] = "redeemOsToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x441a3e70"] = "withdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x46c67b6d"] = "megaSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x46e04a2f"] = "claimTokens";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x47e7ef24"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x4957677c"] = "increase_amount";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x4b8a3529"] = "borrow";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x4cc4a27b"] = "fillOrderRFQToWithPermit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x4d447995"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x4d5693ab"] = "depositUnderlying";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x4d99dd16"] = "undelegate";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x4e3c04bd"] = "getRSETHWithETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x4e71d92d"] = "claim";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x4f498c73"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x4faa8a26"] = "depositEtherFor";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x5028bb95"] = "dodoSwapV2ETHToToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x510ccb43"] = "getReward";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x51227094"] = "processOutputOrders";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x51a60b02"] = "withdrawDelegated";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x5243318f"] = "depositWstETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x5358fbda"] = "depositETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x54741d6d"] = "addOperator";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x54840d1a"] = "swapOnUniswap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x54bacd13"] = "externalSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x54e3f31b"] = "SimpleSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x55362f4d"] = "swapTo";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x573ade81"] = "repay";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x58b9d179"] = "swapOnUniswap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x5a3b74b9"] = "setUserUseReserveAsCollateral";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x5ae401dc"] = "multicall";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x5b0fc9c3"] = "setOwner";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x5b34b966"] = "incrementCounter";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x5c19a95c"] = "delegate";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x5ef2c7f0"] = "setSubnodeRecord";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x60ec5216"] = "addNewValidatorDetails";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x627cdcb9"] = "incrementNonce";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x64466805"] = "swapOnZeroXv4";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x657bb113"] = "mintSign";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x65ae14b5"] = "deleverageWithdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x65fc3873"] = "create_lock";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x66792ba1"] = "send";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x67dfd4c9"] = "leave";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x69328dec"] = "withdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x6ab15071"] = "buyVoucher";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x6b088d5c"] = "withdrawRPL";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x6d104216"] = "burn";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x6d9634b7"] = "releaseTokens";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x6e26b9f8"] = "depositRequest";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x6e553f65"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x6fc3f5d5"] = "deploySmartVault";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x70ccbd31"] = "fillOrderRFQToWithPermit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x71d6e892"] = "claimAlk";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x721c6513"] = "requestExit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x72c51c0b"] = "depositETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x73c2ad9c"] = "addToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x745400c9"] = "requestWithdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x74694a2b"] = "register";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x7617b389"] = "mixSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x7951b76f"] = "requestWithdrawalsWstETHWithPermit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x7b0472f0"] = "stake";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x7bde82f2"] = "redeem";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x7c025200"] = "swap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x7d41c86e"] = "requestRedeem";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x7e108d52"] = "initiateWithdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x7f60d338"] = "withdrawSelfApeCoin";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x81033120"] = "swapOnZeroXv2";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x81791788"] = "dodoMutliSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x827481ea"] = "preSaleMint";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8279e760"] = "claimSelfApeCoin";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x832fbb29"] = "transferFromOrMint";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x833947fd"] = "reallocateTo";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8340f549"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x84bd6d29"] = "clipperSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x852a12e3"] = "redeemUnderlying";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x85f6d155"] = "register";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8697d2c2"] = "claimExitedAssets";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x87201b41"] = "fulfillAvailableAdvancedOrders";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8759c234"] = "unstakeClaimTokens_newPOL";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x891407c0"] = "purchaseTo";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8ab936b8"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8b53f75e"] = "redeemInstant";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8b95dd71"] = "SetAddr";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8bbedf75"] = "ProveAndClaim";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8ca3bf68"] = "callBridgeCall";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8ceab9aa"] = "enterExitQueue";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x8f00eccb"] = "multiSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x90e1aa69"] = "processInputOrders";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x9343d9e1"] = "cooldownShares";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x935fb84b"] = "buyOnUniswap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x94ba89a2"] = "swapBorrowRateMode";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x94bf804d"] = "mint";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x972250fe"] = "bridgeToV2";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x9871efa4"] = "unxswapByOrderId";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x9aa3033a"] = "requestVoluntaryExit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x9dcaafb4"] = "depositSelfApeCoin";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0x9f5db69c"] = "updateOperatorName";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa0712d68"] = "mint";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa1251d75"] = "unoswapWithPermit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa1903eab"] = "submit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa21a9280"] = "withdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa22cb465"] = "setApprovalForAll";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa27e8b6b"] = "undefined";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa2922622"] = "withdrawWithSymbolCheck";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa378534b"] = "create";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa415bcad"] = "borrow";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa543ccea"] = "setWithdrawalAddress";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa5977fbb"] = "send";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa59f3e0c"] = "enter";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa62b2a3d"] = "openPerpetual";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa6886da9"] = "directUniV3Swap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa694fc3a"] = "stake";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa824ae8b"] = "swapFrom";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa8a41c70"] = "cancelOrder_";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa9059cbb"] = "transfer";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xa94e78ef"] = "multiSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xaa2daba6"] = "depositWithSymbolCheck";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xab834bab"] = "atomicMatch_";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xabdb5ea8"] = "repayBorrow";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xac9650d8"] = "Multicall";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xacf1a841"] = "renew";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xacf41e4d"] = "requestWithdrawalsWithPermit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xadcf1163"] = "claim";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xaf68b302"] = "mintToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xafc3083c"] = "removeLiquidity";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb02029e0"] = "extendRewardEmission";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb0431182"] = "clipperSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb1a1a882"] = "depositETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb3a34c4c"] = "fulfillOrder";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb460af94"] = "withdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb6b06dec"] = "requestValidatorsExit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb6b55f25"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb7034f7e"] = "claim";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb7482509"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb7ba18c7"] = "multiClaim";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb80c2f09"] = "smartSwapByOrderId";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xb8aa0db9"] = "depositRsETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xba087652"] = "reedeem";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xbba9b10c"] = "destroy";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xbc80f1a8"] = "uniswapV3SwapTo";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xbd439126"] = "confirmWithdrawalAddress";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xbf509bd4"] = "withdrawELFee";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xbfc2d46a"] = "redeemRequest";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc02dd27a"] = "depositInstant";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc03786b0"] = "buyOnUniswapFork";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc111fb91"] = "preSaleMint";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc22fb4ef"] = "withdrawUnderlying";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc3953502"] = "usePubKeysForValidatorSetup";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc3ae1766"] = "depositAsset";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc47f0027"] = "setName";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc48ef844"] = "createVault";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc4d66de8"] = "initialize";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc5a67b01"] = "initiateWithdrawal";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc5ebeaec"] = "borrow";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc7b8981c"] = "withdrawRewards";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc7cdea37"] = "withdrawETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc805a666"] = "clipperSwapToWithPermit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xc83ec04d"] = "sellVoucher_new";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xca0bfcce"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xcb1c8321"] = "stakeRPLFor";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xcc8690ac"] = "depositCrvLp";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xccc143b8"] = "requestWithdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xcdac52ed"] = "cooldownAssets";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xcfc0afeb"] = "simpleSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xd0a3b665"] = "fillOrderRFQ";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xd0e30db0"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xd1e6044a"] = "compound";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xd3a86833"] = "completeWithdrawal";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xd5f73f5c"] = "redeemFiatRequest";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xd6681042"] = "requestWithdrawals";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xd6a92a5d"] = "clipperSwapToWithPermit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xd7b9d423"] = "withdrawFast";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xd99d32e3"] = "swapAndDeposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xda8be864"] = "undelegate";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xdb006a75"] = "redeem";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xdd307b99"] = "enableOperator";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xddd81f82"] = "registerProxy";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xde0e9a3e"] = "unwrap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xde5f6268"] = "depositAll";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe0db556b"] = "withdrawRewardsPOL";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe1c2eea6"] = "redeemFast";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe2864fe3"] = "cancel";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe2bbb158"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe3afe0a3"] = "claimWithdrawals";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe4457a8a"] = "buyVoucherPOL";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe449022e"] = "uniswapV3Swap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe5234c7a"] = "closePerpetual";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe570b78b"] = "sellVoucher_newPOL";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe61604cf"] = "liquidateBorrow";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe63697c8"] = "withdraw_to_with_slippage";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe69663f1"] = "ZapIn";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe7a050aa"] = "depositIntoStrategy";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe7acab24"] = "fulfillAdvancedOrder";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe7ffb5f7"] = "swapExactAmountIn";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe8a0c121"] = "batchWithdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe8c3516b"] = "getRSETHWithERC20";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe8d6dbb4"] = "RenewAll";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe8eda9df"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe8f28a6c"] = "updateOperatorRewardAddress";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe97fddc2"] = "unstakeClaimTokens_new";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe99454f5"] = "setWithdrawal";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe99a3f80"] = "matchOrders";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xe9fad8ee"] = "exit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xea598cb0"] = "wrap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xea99c2a6"] = "submit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xec1d21dd"] = "megaSwap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xecfe0521"] = "withdrawCrvLp";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xed98a574"] = "fulfillAvailableOrders";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xeda74e71"] = "deleteActiveValidators";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xeea9064b"] = "delegateTo";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xef5e4682"] = "batchWithdrawELFee";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xefef39a1"] = "purchase";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xeff7a612"] = "increase_unlock_time";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf14fcbc8"] = "commit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf2888dbb"] = "unstake";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf2b9fdb8"] = "supply";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf35af1f8"] = "callBridge";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf3fef3a3"] = "withdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf4de10ac"] = "getStakingRewards";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf4f3b200"] = "withdrawERC20";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf532e86a"] = "submit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf56408ed"] = "disableOperator";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf5661034"] = "swapOnUniswapFork";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf5e3c462"] = "liquidateBorrow";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf6326fb3"] = "depositETH";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf756fa21"] = "completeWithdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf7a16963"] = "registerWithConfig";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf8444436"] = "claimWithdrawal";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf87dc1b7"] = "dodoSwapV2TokenToToken";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf8a70486"] = "claimWithdrawal";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf9355f72"] = "buyOnUniswap";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf95a49eb"] = "buy";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xf9609f08"] = "deposit_all";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xfa15b91b"] = "deposit";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xfb0f3ee1"] = "fulfillBasicOrder";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xfb7e92ea"] = "claim";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xfd3c11a8"] = "withdraw";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xfd9f1e10"] = "cancel";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xfe784eaa"] = "swapTokens";
    ETHEREUM_CLEAR_SIGNED_SELECTORS["0xff3bf066"] = "zapInToPT";
})(ETHEREUM_CLEAR_SIGNED_SELECTORS || (exports.ETHEREUM_CLEAR_SIGNED_SELECTORS = ETHEREUM_CLEAR_SIGNED_SELECTORS = {}));
var ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS;
(function (ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS) {
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x08dc9f42"] = "mint";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x0968f264"] = "withdraw";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x0ffab6c2"] = "batchWithdrawCLFee";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x11413601"] = "stableMintSign";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x1249c58b"] = "mint";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x2ba03a79"] = "withdrawCLFee";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x2c4b04fa"] = "deposit";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x4f498c73"] = "deposit";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x657bb113"] = "mintSign";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x66792ba1"] = "send";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x804b936f"] = "stableMint";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0x827481ea"] = "preSaleMint";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0xa0712d68"] = "mint";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0xb6b06dec"] = "requestValidatorsExit";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0xbf509bd4"] = "withdrawELFee";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0xc111fb91"] = "preSaleMint";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0xd0e30db0"] = "deposit";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0xe8a0c121"] = "batchWithdraw";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0xe99454f5"] = "setWithdrawal";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0xef5e4682"] = "batchWithdrawELFee";
    ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS["0xf39247a9"] = "mintSign";
})(ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS || (exports.ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS = ETHEREUM_GOERLI_CLEAR_SIGNED_SELECTORS = {}));
var ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS;
(function (ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS) {
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x01d523b6"] = "updateStateAndDepositAndMintOsToken";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x066055e0"] = "burnOsToken";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x0968f264"] = "withdraw";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x0dd8dd02"] = "queueWithdrawals";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x0ffab6c2"] = "batchWithdrawCLFee";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x18f72950"] = "updateStateAndDeposit";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x1a7ff553"] = "updateState";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x201b9eb5"] = "mintOsToken";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x2ba03a79"] = "withdrawCLFee";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x2c4b04fa"] = "deposit";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x33404396"] = "completeQueuedWithdrawals";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x3a4b66f1"] = "stake";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x4f498c73"] = "deposit";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x66792ba1"] = "send";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x721c6513"] = "requestExit";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x8697d2c2"] = "claimExitedAssets";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x8ceab9aa"] = "enterExitQueue";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0x9aa3033a"] = "requestVoluntaryExit";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xac9650d8"] = "multicall";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xadcf1163"] = "claim";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xb6b06dec"] = "requestValidatorsExit";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xb7ba18c7"] = "multiClaim";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xbf509bd4"] = "withdrawELFee";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xca0bfcce"] = "deposit";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xd0e30db0"] = "deposit";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xda8be864"] = "undelegate";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xe7a050aa"] = "depositIntoStrategy";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xe8a0c121"] = "batchWithdraw";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xe99454f5"] = "setWithdrawal";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xeea9064b"] = "delegateTo";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xef5e4682"] = "batchWithdrawELFee";
    ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS["0xf9609f08"] = "deposit";
})(ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS || (exports.ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS = ETHEREUM_HOLESKY_CLEAR_SIGNED_SELECTORS = {}));
var ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS;
(function (ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS) {
    ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS["0x1249c58b"] = "mint";
    ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS["0x22a775b6"] = "mintAndTransfer";
    ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS["0x27050d1f"] = "createToken";
    ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS["0x3db397c6"] = "__ERC721Rarible_init";
    ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS["0x832fbb29"] = "transferFromOrMint";
    ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS["0xa22cb465"] = "setApprovalForAll";
    ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS["0xc111fb91"] = "preSaleMint";
    ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS["0xe2864fe3"] = "cancel";
    ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS["0xe99a3f80"] = "matchOrders";
})(ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS || (exports.ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS = ETHEREUM_ROPSTEN_CLEAR_SIGNED_SELECTORS = {}));
var FANTOM_CLEAR_SIGNED_SELECTORS;
(function (FANTOM_CLEAR_SIGNED_SELECTORS) {
    FANTOM_CLEAR_SIGNED_SELECTORS["0x00f714ce"] = "withdraw";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x016cba5f"] = "execute";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x0502b1c5"] = "unoswap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x0b86a4c1"] = "swapOnUniswapV2Fork";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x0d5f0e3b"] = "uniswapV3SwapTo";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x12a0ddc7"] = "PMMV2Swap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x12aa3caf"] = "swap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x2298207a"] = "simpleBuy";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x2521b930"] = "uniswapV3SwapToWithPermit";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x2e1a7d4d"] = "withdraw";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x2e95b6c8"] = "unoswap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x3c15fd91"] = "unoswapToWithPermit";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x3ccfd60b"] = "withdraw";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x3d21e25a"] = "swapBridgeToV2";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x3eca9c0a"] = "fillOrderRFQ";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x46c67b6d"] = "megaSwap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x4cc4a27b"] = "fillOrderRFQToWithPermit";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x54840d1a"] = "swapOnUniswap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x54e3f31b"] = "SimpleSwap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x64466805"] = "swapOnZeroXv4";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x6e553f65"] = "deposit";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x70ccbd31"] = "fillOrderRFQToWithPermit";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x7c025200"] = "swap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x81033120"] = "swapOnZeroXv2";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x84bd6d29"] = "clipperSwap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x935fb84b"] = "buyOnUniswap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x972250fe"] = "bridgeToV2";
    FANTOM_CLEAR_SIGNED_SELECTORS["0x9871efa4"] = "unxswapByOrderId";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xa1251d75"] = "unoswapWithPermit";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xa6886da9"] = "directUniV3Swap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xa94e78ef"] = "multiSwap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xb6b55f25"] = "deposit";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xb80c2f09"] = "smartSwapByOrderId";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xbc80f1a8"] = "uniswapV3SwapTo";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xc03786b0"] = "buyOnUniswapFork";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xc805a666"] = "clipperSwapToWithPermit";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xd0a3b665"] = "fillOrderRFQ";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xd0e30db0"] = "deposit";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xe449022e"] = "uniswapV3Swap";
    FANTOM_CLEAR_SIGNED_SELECTORS["0xf5661034"] = "swapOnUniswapFork";
})(FANTOM_CLEAR_SIGNED_SELECTORS || (exports.FANTOM_CLEAR_SIGNED_SELECTORS = FANTOM_CLEAR_SIGNED_SELECTORS = {}));
var OPTIMISM_CLEAR_SIGNED_SELECTORS;
(function (OPTIMISM_CLEAR_SIGNED_SELECTORS) {
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x00f714ce"] = "withdraw";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x016cba5f"] = "execute";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x0502b1c5"] = "unoswap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x0b86a4c1"] = "swapOnUniswapV2Fork";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x0d5f0e3b"] = "uniswapV3SwapTo";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x12a0ddc7"] = "PMMV2Swap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x12aa3caf"] = "swap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x2298207a"] = "simpleBuy";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x259f8d1a"] = "withdraw";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x2e1a7d4d"] = "withdraw";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x2e95b6c8"] = "unoswap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x3c15fd91"] = "unoswapToWithPermit";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x3ccfd60b"] = "withdraw";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x3d21e25a"] = "swapBridgeToV2";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x3eca9c0a"] = "fillOrderRFQ";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x46c67b6d"] = "megaSwap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x4d447995"] = "deposit";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x54840d1a"] = "swapOnUniswap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x54e3f31b"] = "SimpleSwap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x617ba037"] = "supply";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x64466805"] = "swapOnZeroXv4";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x65ae14b5"] = "deleverageWithdraw";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x69328dec"] = "withdraw";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x6e553f65"] = "deposit";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x70ccbd31"] = "fillOrderRFQToWithPermit";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x7c025200"] = "swap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x81033120"] = "swapOnZeroXv2";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x84bd6d29"] = "clipperSwap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x935fb84b"] = "buyOnUniswap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x972250fe"] = "bridgeToV2";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0x9871efa4"] = "unxswapByOrderId";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0xa6886da9"] = "directUniV3Swap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0xa94e78ef"] = "multiSwap";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0xb460af94"] = "withdraw";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0xb6b55f25"] = "deposit";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0xb80c2f09"] = "smartSwapByOrderId";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0xc03786b0"] = "buyOnUniswapFork";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0xc805a666"] = "clipperSwapToWithPermit";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0xd0e30db0"] = "deposit";
    OPTIMISM_CLEAR_SIGNED_SELECTORS["0xf5661034"] = "swapOnUniswapFork";
})(OPTIMISM_CLEAR_SIGNED_SELECTORS || (exports.OPTIMISM_CLEAR_SIGNED_SELECTORS = OPTIMISM_CLEAR_SIGNED_SELECTORS = {}));
var POLYGON_CLEAR_SIGNED_SELECTORS;
(function (POLYGON_CLEAR_SIGNED_SELECTORS) {
    POLYGON_CLEAR_SIGNED_SELECTORS["0x016cba5f"] = "execute";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x02751cec"] = "removeLiquidityETH";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x03388b4e"] = "sublet";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x0502b1c5"] = "unoswap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x0863b7ac"] = "undefined";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x0c65b39d"] = "unlink";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x0d5f0e3b"] = "uniswapV3SwapTo";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x1157decb"] = "endRental";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x11bcc81e"] = "downgrade";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x12a0ddc7"] = "PMMV2Swap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x12aa3caf"] = "swap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x160e8be3"] = "downgradeToETH";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x16dc3ace"] = "swapExactAmountOut";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x18cbafe5"] = "swapExactTokensForETH";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x1a674419"] = "buyCAPSA";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x1ae50184"] = "preSignRentalOffer";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x1beddfdd"] = "endSublet";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x1e6d24c2"] = "dodoSwapV2TokenToETH";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x1e9a6950"] = "redeem";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x20905506"] = "whitelist_profile_creator";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x2195995c"] = "removeLiquidityWithPermit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x2298207a"] = "simpleBuy";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x22a775b6"] = "mintAndTransfer";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x23463624"] = "deposit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x23b872dd"] = "transferFrom";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x2521b930"] = "uniswapV3SwapToWithPermit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x258d3c33"] = "redeemYield";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x26ef699d"] = "sendToken";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x27050d1f"] = "createToken";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x2d8918d7"] = "addLiquidity";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x2e1a7d4d"] = "withdraw";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x2e95b6c8"] = "unoswap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x33635226"] = "undefined";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x38ed1739"] = "swapExactTokensForTokens";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x39255d5b"] = "callAgreement";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x3c15fd91"] = "unoswapToWithPermit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x3c659741"] = "bridgeCall";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x3ce1108d"] = "sellToAMM";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x3d18b912"] = "getReward";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x3d21e25a"] = "swapBridgeToV2";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x3db397c6"] = "__ERC721Rarible_init";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x3eca9c0a"] = "fillOrderRFQ";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x3f2e5fc3"] = "sendNative";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x42966c68"] = "burn";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x441a3e70"] = "withdraw";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x45977d03"] = "upgrade";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x46c67b6d"] = "megaSwap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x4727ce3a"] = "collect";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x47e7ef24"] = "deposit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x48eaf6d6"] = "requestMaticXSwap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x4926c4ed"] = "change_del_exec_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x4a25d94a"] = "swapTokensForExactETH";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x4b7312a9"] = "follow";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x4cc4a27b"] = "fillOrderRFQToWithPermit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x4d7b35a0"] = "follow_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x5028bb95"] = "dodoSwapV2ETHToToken";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x51227094"] = "processOutputOrders";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x51c3659c"] = "change_del_exec_1";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x54506e92"] = "grind";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x54840d1a"] = "swapOnUniswap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x54bacd13"] = "externalSwap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x54e3f31b"] = "SimpleSwap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x560a4db1"] = "act";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x58b9d179"] = "swapOnUniswap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x5b0d5984"] = "removeLiquidityETHWithPermitSupportingFeeOnTransferTokens";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x5b34b966"] = "incrementCounter";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x5c11d795"] = "swapExactTokensForTokensSupportingFeeOnTransferTokens";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x617ba037"] = "supply";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x627cdcb9"] = "incrementNonce";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x6363e882"] = "registerAndRewardGameWithSignature";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x64466805"] = "swapOnZeroXv4";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x65f29f27"] = "quote_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x66b0dcd3"] = "post";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x69328dec"] = "withdraw";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x6ad3ca7d"] = "batchCall";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x6d0fd658"] = "metadata_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x6d9634b7"] = "releaseTokens";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x6dea40b3"] = "set_follow_mod";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x6e09f420"] = "cancelRentalOffer";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x70ccbd31"] = "fillOrderRFQToWithPermit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x7617b389"] = "mixSwap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x77baf209"] = "claimMaticXSwap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x791ac947"] = "swapExactTokensForETHSupportingFeeOnTransferTokens";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x7c025200"] = "swap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x7e502fe0"] = "link_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x7ff36ab5"] = "swapExactETHForTokens";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x809d8947"] = "unfollow_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x81033120"] = "swapOnZeroXv2";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x815ed04a"] = "unfollow";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x81791788"] = "dodoMutliSwap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x825ab164"] = "unlink_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x832fbb29"] = "transferFromOrMint";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x84bd6d29"] = "clipperSwap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x85149258"] = "exerciseOptions";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x86cf48e7"] = "link";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x87201b41"] = "fulfillAvailableAdvancedOrders";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x8803dbee"] = "swapTokensForExactTokens";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x8875eb84"] = "buyFromAMM";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x8ca3bf68"] = "callBridgeCall";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x8d259eb1"] = "rent";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x8f00eccb"] = "undefined";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x907cd7d2"] = "post_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x90e1aa69"] = "processInputOrders";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x935fb84b"] = "buyOnUniswap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x972250fe"] = "bridgeToV2";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x9871efa4"] = "unxswapByOrderId";
    POLYGON_CLEAR_SIGNED_SELECTORS["0x9f916c9f"] = "buyFromAMMWithReferrer";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa1251d75"] = "unoswapWithPermit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa21a9280"] = "withdraw";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa22cb465"] = "setApprovalForAll";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa27e8b6b"] = "undefined";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa2922622"] = "withdrawWithSymbolCheck";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa378534b"] = "create";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa5977fbb"] = "send";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa6886da9"] = "directUniV3Swap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa694fc3a"] = "stake";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa8a41c70"] = "cancelOrder_";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xa94e78ef"] = "multiSwap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xaa2daba6"] = "depositWithSymbolCheck";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xaf2979eb"] = "removeLiquidityETHSupportingFeeOnTransferTokens";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xafc3083c"] = "removeLiquidity";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xb273b4a7"] = "comment";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xb3a34c4c"] = "fulfillOrder";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xb42df51a"] = "comment_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xb60099b7"] = "craft";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xb6b55f25"] = "deposit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xb6f9de95"] = "swapExactETHForTokensSupportingFeeOnTransferTokens";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xb7034f7e"] = "claim";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xb7902e73"] = "collect_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xb80c2f09"] = "smartSwapByOrderId";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xbaa2abde"] = "removeLiquidity";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xbba9b10c"] = "destroy";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xbc80f1a8"] = "uniswapV3SwapTo";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xc03786b0"] = "buyOnUniswapFork";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xc1f4b40a"] = "change_del_exec_2";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xc5d5d96a"] = "act";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xc6d45944"] = "set_block_status_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xc78cf1a0"] = "swapMaticForMaticXViaInstantPool";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xc805a666"] = "clipperSwapToWithPermit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xc86642c7"] = "act_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xce30bb4f"] = "set_follow_mod_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xcf81464b"] = "upgradeByETH";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xcfc0afeb"] = "simpleSwap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xd06a750c"] = "exercisePositions";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xd0a3b665"] = "fillOrderRFQ";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xd0e30db0"] = "deposit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xde5f6268"] = "depositAll";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xded9382a"] = "removeLiquidityETHWithPermit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xe1649e8d"] = "endRentalPrematurely";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xe2864fe3"] = "cancel";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xe2bbb158"] = "deposit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xe449022e"] = "uniswapV3Swap";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xe6a402b5"] = "mirror_with_sign";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xe7acab24"] = "fulfillAdvancedOrder";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xe7ffb5f7"] = "swapExactAmountIn";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xe8e33700"] = "addLiquidity";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xe99a3f80"] = "matchOrders";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xe9fad8ee"] = "exit";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xea25e176"] = "claim";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xec1d21dd"] = "undefined";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xed98a574"] = "fulfillAvailableOrders";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xefe4fd83"] = "metadata";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf08e8f5e"] = "mint";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf0ba35f0"] = "quote";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf2b9fdb8"] = "supply";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf305d719"] = "addLiquidityETH";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf35af1f8"] = "callBridge";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf3fef3a3"] = "withdraw";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf5661034"] = "swapOnUniswapFork";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf87dc1b7"] = "dodoSwapV2TokenToToken";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf90604d1"] = "mirror";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf9355f72"] = "undefined";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xf95a49eb"] = "undefined";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xfb0f3ee1"] = "fulfillBasicOrder";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xfb3bdb41"] = "swapETHForExactTokens";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xfb827e82"] = "set_block_status";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xfd9f1e10"] = "cancel";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xfe9d8323"] = "sellCAPSA";
    POLYGON_CLEAR_SIGNED_SELECTORS["0xff3bf066"] = "zapInToPT";
})(POLYGON_CLEAR_SIGNED_SELECTORS || (exports.POLYGON_CLEAR_SIGNED_SELECTORS = POLYGON_CLEAR_SIGNED_SELECTORS = {}));
var POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS;
(function (POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS) {
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x016cba5f"] = "execute";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x0b86a4c1"] = "swapOnUniswapV2Fork";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x0d5f0e3b"] = "uniswapV3SwapTo";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x12a0ddc7"] = "PMMV2Swap";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x2298207a"] = "simpleBuy";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x3d21e25a"] = "swapBridgeToV2";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x46c67b6d"] = "megaSwap";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x54840d1a"] = "swapOnUniswap";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x54e3f31b"] = "SimpleSwap";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x64466805"] = "swapOnZeroXv4";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x81033120"] = "swapOnZeroXv2";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x935fb84b"] = "buyOnUniswap";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x972250fe"] = "bridgeToV2";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0x9871efa4"] = "unxswapByOrderId";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0xa6886da9"] = "directUniV3Swap";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0xa94e78ef"] = "multiSwap";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0xb80c2f09"] = "smartSwapByOrderId";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0xc03786b0"] = "buyOnUniswapFork";
    POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS["0xf5661034"] = "swapOnUniswapFork";
})(POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS || (exports.POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS = POLYGON_ZK_EVM_CLEAR_SIGNED_SELECTORS = {}));
var ZKSYNC_CLEAR_SIGNED_SELECTORS;
(function (ZKSYNC_CLEAR_SIGNED_SELECTORS) {
    ZKSYNC_CLEAR_SIGNED_SELECTORS["0x0502b1c5"] = "unoswap";
    ZKSYNC_CLEAR_SIGNED_SELECTORS["0x12aa3caf"] = "swap";
    ZKSYNC_CLEAR_SIGNED_SELECTORS["0x3c15fd91"] = "unoswapToWithPermit";
    ZKSYNC_CLEAR_SIGNED_SELECTORS["0x3eca9c0a"] = "fillOrderRFQ";
    ZKSYNC_CLEAR_SIGNED_SELECTORS["0x70ccbd31"] = "fillOrderRFQToWithPermit";
    ZKSYNC_CLEAR_SIGNED_SELECTORS["0x84bd6d29"] = "clipperSwap";
    ZKSYNC_CLEAR_SIGNED_SELECTORS["0xc805a666"] = "clipperSwapToWithPermit";
})(ZKSYNC_CLEAR_SIGNED_SELECTORS || (exports.ZKSYNC_CLEAR_SIGNED_SELECTORS = ZKSYNC_CLEAR_SIGNED_SELECTORS = {}));
